{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0927039",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d5fe0e0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5004f57c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6bfe573f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# IMPORTNAT  this code accept only .fa file as proteome\n",
    "\n",
    "def parse_oma_db(oma_database_address):\n",
    "    \n",
    "    ############### Parsing OMA db ####################\n",
    "    ###################################################\n",
    "\n",
    "    oma_db = db.Database(oma_database_address)\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- OMA data is parsed and its release name is:\", oma_db.get_release_name())\n",
    "    list_speices= [z.uniprot_species_code for z in oma_db.tax.genomes.values()] \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- There are\",len(list_speices),\"species in the OMA database.\")\n",
    "    \n",
    "    return (oma_db, list_speices)\n",
    "\n",
    "\n",
    "def parse_hog_og_map(hog_og_map_address):\n",
    "    \n",
    "    ############### Parsing hog map python dic #############\n",
    "    #########################################################\n",
    "\n",
    "    oma_db = db.Database(oma_database_address)\n",
    "\n",
    "    file = open(hog_og_map_address, \"r\")\n",
    "    contents = file.read()\n",
    "    hog_OG_map = ast.literal_eval(contents)\n",
    "    file.close()\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- The hog-og map is read from file with the length of \", len(hog_OG_map))\n",
    "       \n",
    "    return (hog_OG_map)\n",
    "\n",
    "\n",
    "\n",
    "def HOG_OG_map(oma_database_address):\n",
    "\n",
    "    oma_db = db.Database(oma_database_address)\n",
    "    print(\"OMA data is parsed and its release name is:\", oma_db.get_release_name())\n",
    "    list_speices= [z.uniprot_species_code for z in oma_db.tax.genomes.values()] \n",
    "    print(\"There are\",len(list_speices),\"species in the OMA database.\")\n",
    "\n",
    "    hog_OG_map = {hog_id.decode(\"utf-8\") : og_count_list[0][0] for hog_id, og_count_list in build_hog_to_og_map(oma_db)}\n",
    "    print(\"HOG-OG map is extracted with length of \", len(hog_OG_map))\n",
    "\n",
    "\n",
    "    hog_OG_map_address = oma_database_address+\"_hog_og_map.dic\"\n",
    "    with open(hog_OG_map_address, 'w') as file_map:\n",
    "        print(hog_OG_map, file=file_map)\n",
    "\n",
    "    file_map.close()\n",
    "    print(\"HOG-OG map is written in the file\",hog_OG_map_address)\n",
    "    \n",
    "    return hog_OG_map\n",
    "\n",
    "\n",
    "def parse_proteome(list_speices):\n",
    "    \n",
    "    ############### Parsing query proteome of species #######\n",
    "    #########################################################\n",
    "\n",
    "    \n",
    "    # IMPORTNAT  this code accept only .fa file as proteome\n",
    "    project_files = listdir(project_folder+\"/omamer_search/proteome/\")\n",
    "\n",
    "    query_species_names = []\n",
    "    for file in project_files:\n",
    "        if file.split(\".\")[-1]==\"fa\":\n",
    "            file_name_split = file.split(\".\")[:-1]\n",
    "            query_species_names.append('.'.join(file_name_split))\n",
    "        if file.split(\".\")[-1]==\"fasta\":\n",
    "            file_name_split = file.split(\".\")[:-1]\n",
    "            query_species_names.append('.'.join(file_name_split))\n",
    "\n",
    "    # we may assert existence of query_species_name+\".fa/hogmap\"\n",
    "    query_prot_records_species = [ ]\n",
    "    for query_species_name in query_species_names:\n",
    "        query_prot_address = project_folder +\"omamer_search/proteome/\"+ query_species_name + \".fa\" \n",
    "        \n",
    "        query_prot_records = list(SeqIO.parse(query_prot_address, \"fasta\")) \n",
    "        query_prot_records_species.append(query_prot_records)\n",
    "\n",
    "    query_species_num = len(query_species_names)    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- The are\",str(query_species_num),\"species in the proteome folder.\")\n",
    "\n",
    "    # for development\n",
    "    query_species_num = len(query_species_names)\n",
    "    for species_i in range(query_species_num):\n",
    "        len_prot_record_i = len( query_prot_records_species[species_i] )\n",
    "        species_name_i = query_species_names[species_i]\n",
    "        #print(species_name_i,len_prot_record_i)\n",
    "        if species_name_i in list_speices: \n",
    "            current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "            print(current_time,\"- the species\",species_name_i,\" already exists in the oma database, remove them first\")\n",
    "            exit()\n",
    "\n",
    "      \n",
    "    # The proteins are parsed using  Bio.SeqIO.parse\n",
    "    # the first part of the header line before space \n",
    "    # >tr|A0A2I3FYY2|A0A2I3FYY2_NOMLE Uncharacterized protein OS=Nomascus leucogenys OX=61853 GN=CLPTM1L PE=3 SV=1\n",
    "    # will be \">tr|A0A2I3FYY2|A0A2I3FYY2_NOMLE\"\n",
    "    # [i.id for i in query_prot_records_species[0] if len(i.id)!=30 and len(i.id)!=22 ] #'sp|O47892|CYB_NOMLE',\n",
    "    \n",
    "\n",
    "    \n",
    "    return (query_species_names, query_prot_records_species)\n",
    "\n",
    "\n",
    "\n",
    "def parse_hogmap_omamer(query_species_names):\n",
    "\n",
    "    ################### Parsing omamer's output  ########\n",
    "    #####################################################\n",
    "\n",
    "    query_prot_names_species_mapped = []\n",
    "    query_hogids_species = []\n",
    "\n",
    "    for query_species_name in query_species_names:\n",
    "        omamer_output_address = project_folder + \"omamer_search/hogmap/\"+ query_species_name + \".hogmap\"     \n",
    "        omamer_output_file = open(omamer_output_address,'r');\n",
    "\n",
    "        query_prot_names= []\n",
    "        query_hogids= []\n",
    "\n",
    "        for line in omamer_output_file:\n",
    "            line_strip=line.strip()\n",
    "            if not line_strip.startswith('qs'):\n",
    "                line_split= line_strip.split(\"\\t\")    \n",
    "                #if line_split[1]!='na':\n",
    "                query_prot_names.append(line_split[0])\n",
    "                query_hogids.append(line_split[1])\n",
    "        #print(\"number of proteins in omamer output for \",query_species_name,\"is\",len(query_hogids)) # ,query_hogids\n",
    "        query_prot_names_species_mapped.append(query_prot_names)\n",
    "        query_hogids_species.append(query_hogids)    \n",
    "        \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- There are \",len(query_prot_names_species_mapped),\" in hogmap folder.\")\n",
    "    print(current_time,\"- The first species\",query_species_names[0],\" contains \",len(query_hogids_species[0]),\" proteins.\")\n",
    "    print(current_time,\"- The first protein of first species is \",query_prot_names[0])\n",
    "\n",
    "    \n",
    "    return (query_prot_names_species_mapped, query_hogids_species)\n",
    "    \n",
    "    \n",
    "\n",
    "def filter_prot_mapped(query_species_names, query_prot_records_species,query_prot_names_species_mapped):\n",
    "    # omamer remove very small proteins, \n",
    "    # so  we lose track of order comparing hogmap and fasta file\n",
    "    # the goal here is to remove those from seq record (of the fasta file)\n",
    "    query_prot_records_species_filtered=[]\n",
    "    for species_idx in range(len(query_species_names)):    \n",
    "        # from fasta file\n",
    "        query_species_name=query_species_names[species_idx]\n",
    "        query_prot_records_species_i = query_prot_records_species[species_idx]\n",
    "        query_prot_ids_records = [record.id for record in query_prot_records_species_i]\n",
    "\n",
    "        # from hogmap file\n",
    "        # without proteins that are not mapped on any hogs\n",
    "        query_prot_names_species_i = query_prot_names_species_mapped[species_idx]\n",
    "\n",
    "        if len(query_prot_names_species_i) != len(query_prot_records_species_i):\n",
    "\n",
    "            query_prot_records_filterd=[]\n",
    "            for query_prot_name in query_prot_names_species_i:\n",
    "                if query_prot_name in query_prot_ids_records:\n",
    "                    prot_record_idx = query_prot_ids_records.index(query_prot_name)\n",
    "                    prot_record = query_prot_records_species_i[prot_record_idx]\n",
    "                    query_prot_records_filterd.append(prot_record)\n",
    "                else:\n",
    "                    current_time = datetime.now().strftime(\"%H:%M:%S\")                    \n",
    "                    print(current_time,\"- Error\",query_species_name, query_prot_name)\n",
    "\n",
    "            current_time = datetime.now().strftime(\"%H:%M:%S\")        \n",
    "            print(current_time,\"- For the species\", query_species_name, \", few proteins were ignored by omamer.\")\n",
    "            print(current_time,\"- before filtering: in hogmap\", len(query_prot_names_species_i), \"in proteome\", len(query_prot_records_species_i))\n",
    "            print(current_time,\"- After filtering:  in hogmap\", len(query_prot_names_species_i), \"in proteome\", len(query_prot_records_filterd))            \n",
    "            \n",
    "\n",
    "        else:\n",
    "            query_prot_records_filterd = query_prot_records_species_i\n",
    "\n",
    "        query_prot_records_species_filtered.append(query_prot_records_filterd)\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")        \n",
    "    print(current_time,\"- For the rest of species, all proteins were mapped using OMAmer.\")\n",
    "\n",
    "    return query_prot_records_species_filtered\n",
    "\n",
    "\n",
    "\n",
    "def extract_unique_hog_pure(query_species_names,query_hogids_species, query_prot_names_species_mapped,query_prot_records_species):\n",
    "    ###### Extracting unique HOG list and corresponding query proteins ########\n",
    "    ###########################################################################\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- Extracting proteins mapped only once on a HOG is started\")\n",
    "\n",
    "    query_hogids_filtr_species = []\n",
    "    query_prot_names_filtr_species = []\n",
    "    query_prot_records_filtr_species = []\n",
    "\n",
    "    repeated_hogid_num = 0\n",
    "    \n",
    "    query_species_num = len(query_species_names) \n",
    "    \n",
    "    dic_hogs_list=[]  # a list of dictinaries \n",
    "\n",
    "\n",
    "    for species_i in range(query_species_num):\n",
    "\n",
    "        query_hogids =  query_hogids_species[species_i]\n",
    "        query_prot_names = query_prot_names_species_mapped[species_i]\n",
    "        query_prot_records  = query_prot_records_species[species_i]\n",
    "        \n",
    "        dic_hogs = {}\n",
    "        for prot_i in range(len(query_hogids)):\n",
    "            query_hogid      = query_hogids[prot_i]\n",
    "            query_prot_name  = query_prot_names[prot_i]\n",
    "            query_prot_record= query_prot_records[prot_i]\n",
    "            if query_hogid!= 'na':\n",
    "                if  query_hogid  not in  dic_hogs:\n",
    "                    dic_hogs[query_hogid]=[(query_prot_name,query_prot_record)]\n",
    "                else:\n",
    "                    repeated_hogid_num += 1 \n",
    "                    dic_hogs[query_hogid].append((query_prot_name,query_prot_record))\n",
    "        dic_hogs_list.append(dic_hogs)\n",
    "\n",
    "        \n",
    "    for dic_hogs in dic_hogs_list: # each species\n",
    "        \n",
    "        query_prot_names_filtr = []\n",
    "        query_prot_records_filtr = []\n",
    "\n",
    "        hogid_list = list(dic_hogs.keys())\n",
    "        for hogid in hogid_list:\n",
    "            list_query_prot = dic_hogs[hogid]\n",
    "            if len(list_query_prot)>1:\n",
    "                del dic_hogs[hogid]\n",
    "        #here dic_hogs is updated and  dic_hogs_list  is also updated.\n",
    "        \n",
    "        #print(len(hogid_list),len(dic_hogs))\n",
    "        #for key, value in d.items():\n",
    "        query_hogids_filtr = []\n",
    "        query_prot_names_filtr = []\n",
    "        query_prot_records_filtr = []        \n",
    "        #query_hogids_filtr=list(dic_hogs.keys())\n",
    "        for hogid, tuple_value  in dic_hogs.items():\n",
    "            query_hogids_filtr.append(hogid)\n",
    "            query_prot_names_filtr.append(tuple_value[0][0])\n",
    "            query_prot_records_filtr.append(tuple_value[0][1])\n",
    "        \n",
    "        \n",
    "        \n",
    "\n",
    "        query_hogids_filtr_species.append(query_hogids_filtr)\n",
    "        query_prot_names_filtr_species.append(query_prot_names_filtr)\n",
    "        query_prot_records_filtr_species.append(query_prot_records_filtr)        \n",
    "    \n",
    "    current_time  = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- Extracting proteins mapped only once on a HOG is finished\")\n",
    "    num_all_hogs=np.sum([len(dic_hogs) for dic_hogs in dic_hogs_list])\n",
    "    print(current_time,\"- For \",len(dic_hogs_list),\" species, we keep \",num_all_hogs,\"HOGs.\")\n",
    "    \n",
    "\n",
    "    return (query_hogids_filtr_species, query_prot_names_filtr_species, query_prot_records_filtr_species )\n",
    "  \n",
    "\n",
    "def gather_OG(query_species_names, query_hogids_filtr_species, query_prot_names_filtr_species, query_prot_records_filtr_species):\n",
    "\n",
    "    ############ Extracting the most frequent OG  ########\n",
    "    #####################################################\n",
    "\n",
    "    #dict (oma_group_nr -> dict(species, [proteins]))\n",
    "    #Og[555] = {homo_erectus: [blabla, blublu], yellow_bird: [P52134], brown_bear: [P2121,B53223]}\n",
    "\n",
    "    OGs_queries = {}\n",
    "\n",
    "    # hog_OG_map = {}\n",
    "\n",
    "    mostFrequent_OG_list_species = []\n",
    "\n",
    "    frq_most_frequent_og_list_all = [] # for development\n",
    "    \n",
    "    query_species_num = len(query_species_names)  \n",
    "    for species_i in  range(query_species_num):\n",
    "\n",
    "        query_species_name = query_species_names[species_i]\n",
    "        #print(\"\\n\",query_species_name)\n",
    "\n",
    "        query_hogids_filtr = query_hogids_filtr_species[species_i]\n",
    "        query_prot_names_filtr = query_prot_names_filtr_species[species_i]\n",
    "        query_prot_records_filtr = query_prot_records_filtr_species[species_i]\n",
    "\n",
    "        mostFrequent_OG_list=[]\n",
    "\n",
    "        num_query_filtr = len(query_hogids_filtr)\n",
    "        for  item_idx in range(num_query_filtr): #\n",
    "\n",
    "            #query_protein = query_prot_names_filtr[item_idx]\n",
    "            seqRecords_query =  query_prot_records_filtr[item_idx]\n",
    "            seqRecords_query_edited = SeqRecord(Seq(str(seqRecords_query.seq)), query_species_name, '', '')\n",
    "            #print(seqRecords_query_edited)\n",
    "\n",
    "            hog_id= query_hogids_filtr[item_idx]\n",
    "\n",
    "            if not hog_id in hog_OG_map:   # Caculitng  most frq OG for the new hog\n",
    "                mostFrequent_OG= -1\n",
    "                hog_OG_map[hog_id]=mostFrequent_OG\n",
    "\n",
    "            else:  # hog_id is in hog_OG_map dic\n",
    "                #print(\"using the hog-og-map\")\n",
    "                mostFrequent_OG = hog_OG_map[hog_id]\n",
    "\n",
    "            if mostFrequent_OG in OGs_queries:\n",
    "                OGs_queries_k = OGs_queries[mostFrequent_OG]\n",
    "\n",
    "                if not query_species_name in OGs_queries_k:\n",
    "                    OGs_queries_k[query_species_name] = seqRecords_query_edited\n",
    "                    OGs_queries[mostFrequent_OG] = OGs_queries_k\n",
    "            else:\n",
    "                OGs_queries[mostFrequent_OG] = {query_species_name: seqRecords_query_edited} # query_protein = query_prot_names_filtr[item_idx]\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Needed HOH-OG map \",len(OGs_queries),\"are extracted from the map file.\") \n",
    "    \n",
    "    return OGs_queries\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "def filter_gathered_OG(OGs_queries, query_species_names,keep_og_treshold_species_query):\n",
    "\n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Start filtering OGs of proteme queries with length of \", len(OGs_queries),\"for \",len(query_species_names),\"species.\") \n",
    "       \n",
    "        \n",
    "        \n",
    "    OGs_queries_filtr={}\n",
    "\n",
    "    query_species_og_kept_set= set()\n",
    "\n",
    "    query_species_num_OGs_list=[]\n",
    "    for OG_id, values in OGs_queries.items():\n",
    "        query_species_og= list(values.keys())\n",
    "\n",
    "        query_species_num_OGs_list.append(len(query_species_og)) \n",
    "        if len(query_species_og) > keep_og_treshold_species_query:\n",
    "            OGs_queries_filtr[OG_id]=values\n",
    "            query_species_og_kept_set= query_species_og_kept_set.union(set(query_species_og))\n",
    "\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Finished filetering from \",len(OGs_queries),\", now\",len(OGs_queries_filtr),\"OGs are remained.\") \n",
    "               \n",
    "\n",
    "\n",
    "    not_included_species =[]\n",
    "    for i in set(query_species_names):\n",
    "        if i not in query_species_og_kept_set:\n",
    "            not_included_species.append(i)\n",
    "\n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- These species missed due to the threshold of\",keep_og_treshold_species_query,\":\",not_included_species) \n",
    "               \n",
    "\n",
    "\n",
    "    #query_species_num_OGs_list\n",
    "    # for i in not_included_species:\n",
    "    #     #print(i)\n",
    "    #     for OG_id, values in OGs_queries.items():\n",
    "    #         if i in values:\n",
    "    #             print(values.keys())\n",
    "    # easiet way keep the first one or make list and keep the best\n",
    "    # but better to set trehsold not to misss any species \n",
    "    \n",
    "    return OGs_queries_filtr\n",
    "    \n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "def combine_OG_query_filtr(OGs_queries_filtr, oma_db, threshold_least_query_sepecies_in_OG,kept_oma_species_num):\n",
    "    \n",
    "#     ########## Combine proteins of OG with queries ##################\n",
    "#     #################################################################\n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Combining queries with OG started for\",len(OGs_queries_filtr),\"OGs.\") \n",
    "    \n",
    "    species_og_dic={}\n",
    "    \n",
    "    proteins_object_OG_dic={}\n",
    "    for OG_q in OGs_queries_filtr.keys():  # OG found in the query\n",
    "\n",
    "        dic_species_prot = OGs_queries_filtr[OG_q]\n",
    "        if len(dic_species_prot) >threshold_least_query_sepecies_in_OG:\n",
    "            if OG_q != -1:\n",
    "                OG_members = oma_db.oma_group_members(OG_q)\n",
    "                proteins_object_OG = [db.ProteinEntry(oma_db, pr) for pr in OG_members]  # the protein IDs of og members\n",
    "                proteins_object_OG_dic[OG_q]=proteins_object_OG\n",
    "                \n",
    "                species_all_og = [ str(pr.genome.uniprot_species_code) for pr in proteins_object_OG ]\n",
    "\n",
    "                for species in species_all_og:\n",
    "                    if species in species_og_dic:\n",
    "                        #species_og_dic[species].append(OG_q)                        \n",
    "                        species_og_dic[species] +=1 \n",
    "                    else:\n",
    "                        #species_og_dic[species]=[OG_q]\n",
    "                        species_og_dic[species] = 1\n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Number of OMA species per OG is calculated\") \n",
    "                         \n",
    "    species_oma_list=[]\n",
    "    og_num_list = []\n",
    "    for species_oma, og_num in species_og_dic.items():\n",
    "        og_num_list.append(og_num)\n",
    "        species_oma_list.append(species_oma)\n",
    "        \n",
    "    id_species_keep = np.argsort(og_num_list)[-kept_oma_species_num:]\n",
    "    species_oma_kept = [species_oma_list[i] for i in id_species_keep]\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- These OMA species are kept\",species_oma_kept) \n",
    "\n",
    "    \n",
    "    seqRecords_OG_queries = []\n",
    "    seqRecords_all_filtr = []\n",
    "    for OG_q in OGs_queries_filtr.keys():  # OG found in the query\n",
    "\n",
    "        dic_species_prot = OGs_queries_filtr[OG_q]\n",
    "        if len(dic_species_prot) >threshold_least_query_sepecies_in_OG:\n",
    "            \n",
    "            seqRecords_query_edited_all = list(dic_species_prot.values())\n",
    "            if OG_q != -1:\n",
    "                proteins_object_OG = proteins_object_OG_dic[OG_q]\n",
    "                 # covnert to biopython objects\n",
    "                seqRecords_OG = []\n",
    "                for pr in proteins_object_OG:\n",
    "                    species_code = str(pr.genome.uniprot_species_code)\n",
    "                    if species_code in species_oma_kept:\n",
    "                        seq_record = SeqRecord(Seq(pr.sequence),species_code,'','') \n",
    "                        seqRecords_OG.append(seq_record)\n",
    "                \n",
    "                seqRecords_OG_queries =seqRecords_OG + seqRecords_query_edited_all\n",
    "                current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "                #print(current_time, \" - Length of OG\",OG_q,\"was\",len(seqRecords_OG),\",now is\",len(seqRecords_OG_queries))\n",
    "                print(current_time, \" - Combining OG\",OG_q,\" with length of \",len(seqRecords_OG),\"\\t with a query \",len(seqRecords_query_edited_all),\" is just finished.\")\n",
    "\n",
    "                seqRecords_all_filtr.append(seqRecords_OG_queries)\n",
    "\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(\"\\n\", current_time, \"- Combining queries with OG is finished! number of OGs\",len(seqRecords_all_filtr)) # \n",
    "    \n",
    "    \n",
    "    \n",
    "    open_file = open(project_folder+\"fastoma_core/_file_combined_OGs_filtr.pkl\", \"wb\")\n",
    "    pickle.dump(seqRecords_all_filtr, open_file)\n",
    "    open_file.close()\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(\"\\n\", current_time, \"- The variable seqRecords_all_filtr is saved as\", project_folder+\"fastoma_core/_file_combined_OGs_filtr.pkl\") # \n",
    "    \n",
    "    return(seqRecords_all_filtr)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def run_msa_OG(seqRecords_OG_queries):\n",
    "    ############## MSA  ##############\n",
    "    ##################################\n",
    "    #current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    #print(current_time, \"- working on new OG with length of \",len(seqRecords_OG_queries))\n",
    "\n",
    "    wrapper_mafft = mafft.Mafft(seqRecords_OG_queries,datatype=\"PROTEIN\") \n",
    "    # MAfft error: Alphabet 'U' is unknown. -> add --anysymbol argument needed to define in the sourse code\n",
    "    # workaround sed \"s/U/X/g\"\n",
    "    \n",
    "    wrapper_mafft.options.options['--retree'].set_value(1)\n",
    "\n",
    "\n",
    "    run_mafft = wrapper_mafft() # it's wrapper  storing the result  and time \n",
    "    time_taken_mafft = wrapper_mafft.elapsed_time\n",
    "\n",
    "    result_mafft = wrapper_mafft.result \n",
    "    time_taken_mafft2 = wrapper_mafft.elapsed_time\n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    #print(current_time,\"- time elapsed for MSA: \",time_taken_mafft2)\n",
    "    print(current_time,\"- MSA for an OG is just finished: \",time_taken_mafft2)\n",
    "\n",
    "    return(result_mafft)\n",
    "   \n",
    "\n",
    "\n",
    "def run_msa_OG_parallel(seqRecords_all,number_max_workers):\n",
    "        \n",
    "    iterotr_OGs = 0 \n",
    "    \n",
    "    result_mafft_all_species=[]\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Parallel msa is started for \",len(seqRecords_all),\" OGs.\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    with concurrent.futures.ProcessPoolExecutor(max_workers=number_max_workers) as executor: \n",
    "        for seqRecords_OG_queries, output_values in zip(seqRecords_all, executor.map(run_msa_OG, seqRecords_all)):\n",
    "            result_mafft_all_species.append(output_values)\n",
    "\n",
    "            \n",
    "    \n",
    "    open_file = open(project_folder+\"fastoma_core/_file_msas.pkl\", \"wb\")\n",
    "    pickle.dump(result_mafft_all_species, open_file)\n",
    "    open_file.close()\n",
    "    \n",
    "    return result_mafft_all_species\n",
    "    \n",
    "\n",
    "def filter_ogs(result_mafft_all_species,ogs_keep_number):\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Filtering MSA started.\")\n",
    "\n",
    "    density_ogs=[]\n",
    "    for msa_og in result_mafft_all_species:\n",
    "\n",
    "        gap_count_og= 0\n",
    "        all_count_og=0\n",
    "        for record in msa_og:\n",
    "            seq=str(record.seq)\n",
    "            gap_count_og += seq.count(\"-\") + seq.count(\"?\") + seq.count(\".\") +seq.count(\"~\")    \n",
    "        #gap_count_ogs.append(gap_count_og)\n",
    "        density_og=gap_count_og/(len(msa_og)*len(msa_og[0]))\n",
    "        density_ogs.append(density_og)\n",
    "        #if density_ogs> treshold_density:\n",
    "    plt.hist(density_ogs,bins=100) # , bins=10\n",
    "    #plt.show()\n",
    "    plt.savefig(project_folder+\"fastoma_core/_density_ogs.pdf\")\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- The histogram of density ogs is saved.\")\n",
    "    \n",
    "    id_og_keep = np.argsort(density_ogs)[-ogs_keep_number:]\n",
    "    result_mafft_all_species_filtr = [result_mafft_all_species[i] for i in id_og_keep]\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Filtering MSA finished, keeping\",len(result_mafft_all_species_filtr),\"out of\",len(result_mafft_all_species))       \n",
    "    \n",
    "\n",
    "    open_file = open(project_folder+\"fastoma_core/_\"+str(ogs_keep_number)+\"_file_msas_filtered.pkl\", \"wb\")\n",
    "    pickle.dump(result_mafft_all_species_filtr, open_file)\n",
    "    open_file.close()\n",
    "\n",
    "        \n",
    "    return result_mafft_all_species_filtr\n",
    "\n",
    "\n",
    "\n",
    "def concatante_alignments(result_mafft_all_species,ogs_keep_number):\n",
    "    ############## Concatante alignments  ##############\n",
    "    ####################################################\n",
    "\n",
    "    #alignments= result_maf2_all\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- MSA concatenation started\")\n",
    "    \n",
    "    \n",
    "    alignments= result_mafft_all_species\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Alignments len\",len(alignments))\n",
    "    #print([len(aln) for aln in alignments ])\n",
    "    #print([len(seq) for aln in alignments for seq in aln])\n",
    "\n",
    "    all_labels_raw = [seq.id for aln in alignments for seq in aln]\n",
    "    all_labels = set(all_labels_raw)\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- ids: \",len(all_labels),len(all_labels_raw))\n",
    "    \n",
    "    # Make a dictionary to store info as we go along\n",
    "    # (defaultdict is convenient -- asking for a missing key gives back an empty list)\n",
    "    concat_buf = defaultdict(list)\n",
    "\n",
    "    # Assume all alignments have same alphabet\n",
    "    alphabet = alignments[0]._alphabet\n",
    "\n",
    "    for aln in alignments:\n",
    "        length = aln.get_alignment_length()\n",
    "        #print(\"length\",length)\n",
    "        # check if any labels are missing in the current alignment\n",
    "        these_labels = set(rec.id for rec in aln)\n",
    "        missing = all_labels - these_labels\n",
    "        #print(missing)\n",
    "        # if any are missing, create unknown data of the right length,\n",
    "        # stuff the string representation into the concat_buf dict\n",
    "        for label in missing:\n",
    "            new_seq = UnknownSeq(length, alphabet=alphabet)\n",
    "            concat_buf[label].append(str(new_seq))\n",
    "\n",
    "        # else stuff the string representation into the concat_buf dict\n",
    "        for rec in aln:\n",
    "            concat_buf[rec.id].append(str(rec.seq))\n",
    "\n",
    "    # Stitch all the substrings together using join (most efficient way),\n",
    "    # and build the Biopython data structures Seq, SeqRecord and MultipleSeqAlignment\n",
    "    msa = MultipleSeqAlignment(SeqRecord(Seq(''.join(seq_arr), alphabet=alphabet), id=label)\n",
    "                                for (label, seq_arr) in concat_buf.items())\n",
    "\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- MSA concatenation finished\")\n",
    "\n",
    "    ##########\n",
    "    # print the borders of concatatnion  in nexus format\n",
    "    # http://www.iqtree.org/doc/Advanced-Tutorial\n",
    "    #######\n",
    "    out_name_msa=project_folder+\"fastoma_core/_\"+str(ogs_keep_number)+\"_msa_concatanated.txt\"\n",
    "    handle_msa_fasta = open(out_name_msa,\"w\")\n",
    "    SeqIO.write(msa, handle_msa_fasta,\"fasta\")\n",
    "    handle_msa_fasta.close()\n",
    "\n",
    "\n",
    "    # sequences_all=[]\n",
    "    # for spec, sequences in concat_buf.items():\n",
    "    #      sequences_all.append(sequences)\n",
    "    # for i in range(1,100):\n",
    "    #     print(len(sequences_all[0][i]),len(sequences_all[2][i]) ,len(sequences_all[40][i]) )\n",
    "    ## therefore, all of them are with same length of sequences\n",
    "\n",
    "    species_0=list(concat_buf.keys())[0]\n",
    "    sequences =  concat_buf[species_0]\n",
    "    len_all = [len(i) for i in sequences] \n",
    "#     borders_in_supermatrix=[]\n",
    "#     borders_in_supermatrix.append((1,len_all[0] )) \n",
    "#     accumalte = len_all[0]\n",
    "#     for i in range(1,100):\n",
    "#         borders_in_supermatrix.append((accumalte+1,accumalte+len_all[i]))\n",
    "#         accumalte +=len_all[i]\n",
    "\n",
    "#     print(len(msa),len(msa[0]),accumalte)\n",
    "#     #check it works or not\n",
    "#     # for i in range(100):\n",
    "#     #     print(borders_in_supermatrix[i][1]-borders_in_supermatrix[i][0]+1,len_all[i])\n",
    "\n",
    "#     #borders_in_supermatrix\n",
    "    \n",
    "#     # begin sets;\n",
    "#     #     charset part1 = 1-100;\n",
    "#     #     charset part2 = 101-384;\n",
    "#     #     charpartition mine = HKY+G:part1, GTR+I+G:part2;\n",
    "#     # end;\n",
    "#     borders_out_file_add=project_folder+\"border_out_parition.nex\"\n",
    "#     borders_out_file = open(borders_out_file_add,'w')\n",
    "\n",
    "#     borders_out_file.write(\"#nexus\"+\"\\n\")\n",
    "#     borders_out_file.write(\"begin sets;\"+\"\\n\")\n",
    "\n",
    "#     for i, val in enumerate(borders_in_supermatrix):\n",
    "\n",
    "#         line=\"    charset part\"+str(i)+\" = \"+str(val[0])+\"-\"+str(val[1])+\";\"\n",
    "#         #print(line)\n",
    "#         borders_out_file.write(line+'\\n')\n",
    "\n",
    "#     borders_out_file.write(\"end;\\n\")\n",
    "#     borders_out_file.close()\n",
    "#     print(borders_out_file_add)\n",
    "    \n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- MSA concatenation has been written in the file\",out_name_msa,\"with length\", len(msa),msa.get_alignment_length()) # super matrix size\n",
    "    \n",
    "    \n",
    "    return msa\n",
    "    \n",
    "\n",
    "def msa_filter_row(msa,tresh_ratio_gap_row,query_species_names,ogs_keep_number):\n",
    "\n",
    "    msa_filtered_row = [] # msa_fltr\n",
    "    ratio_records=[]\n",
    "    for record in msa:\n",
    "        species_name = record.id\n",
    "        seq = record.seq\n",
    "        seqLen = len(record)\n",
    "        \n",
    "        gap_count=seq.count(\"-\") + seq.count(\"?\") + seq.count(\".\") +seq.count(\"~\")\n",
    "                \n",
    "        ratio_record_nongap= 1-gap_count/seqLen\n",
    "        ratio_records.append(round(ratio_record_nongap,3))\n",
    "\n",
    "        if ratio_record_nongap > tresh_ratio_gap_row:\n",
    "            msa_filtered_row.append(record)\n",
    "        elif species_name in query_species_names : \n",
    "            msa_filtered_row.append(record)\n",
    "            current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "            print(current_time, \"- Many row-wise gap for query\",species_name,\"with a ratio of\",ratio_record_nongap) \n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "\n",
    "    print(current_time, \"- Row-wise filtering of MSA is finished.\") \n",
    "    print(current_time, \"- Out of \",len(msa),\"species,\",len(msa_filtered_row),\"species (row of msa) remained.\")\n",
    "\n",
    "    out_name_msa=project_folder+\"fastoma_core_\"+str(ogs_keep_number)+\"_msa_concatanated_filtered_row_\"+str(tresh_ratio_gap_row)+\".txt\"\n",
    "    handle_msa_fasta = open(out_name_msa,\"w\")\n",
    "    SeqIO.write(msa_filtered_row, handle_msa_fasta,\"fasta\")\n",
    "    handle_msa_fasta.close()\n",
    "    \n",
    "    print(current_time, \"- MSA Row-wise filtered stored in file.\") # super matrix size\n",
    "    \n",
    "    \n",
    "    return msa_filtered_row\n",
    "    \n",
    "    \n",
    "\n",
    "def msa_filter_col(msa_filtered_row, tresh_ratio_gap_col,tresh_ratio_gap_row,ogs_keep_number):\n",
    "\n",
    "    ratio_col_all = []\n",
    "\n",
    "    length_record= len(msa_filtered_row[1])\n",
    "    num_records = len(msa_filtered_row)\n",
    "\n",
    "\n",
    "    keep_cols = []\n",
    "    for col_i in range(length_record):  # inspired by https://github.com/andreas-wilm/compbio-utils/blob/master/prune_aln_cols.py \n",
    "\n",
    "        col_values = [record.seq[col_i] for record in msa_filtered_row]\n",
    "\n",
    "        gap_count=col_values.count(\"-\") + col_values.count(\"?\") + col_values.count(\".\") +col_values.count(\"~\")\n",
    "\n",
    "        ratio_col_nongap = 1- gap_count/num_records\n",
    "        ratio_col_all.append(ratio_col_nongap)\n",
    "        if ratio_col_nongap  > tresh_ratio_gap_col:\n",
    "            keep_cols.append(col_i)\n",
    "\n",
    "\n",
    "    plt.hist(ratio_col_all,bins=100) # , bins=10\n",
    "    #plt.show()\n",
    "    plt.savefig(project_folder+\"fastoma_core/_ratio_col.pdf\")\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Columns indecis extracted. Out of \", length_record,\"columns,\",len(keep_cols),\"is remained.\") \n",
    "\n",
    "    msa_filtered_row_col = []\n",
    "\n",
    "    for record in msa_filtered_row :\n",
    "        record_seq = str(record.seq)\n",
    "\n",
    "        record_seq_edited  = ''.join([record_seq[i] for i in keep_cols  ])\n",
    "        record_edited= SeqRecord(Seq(record_seq_edited), record.id, '', '')\n",
    "        msa_filtered_row_col.append(record_edited)                         \n",
    "\n",
    "\n",
    "    out_name_msa=project_folder+\"fastoma_core/_\"+str(ogs_keep_number)+\"_msa_concatanated_filtered_row_\"+str(tresh_ratio_gap_row)+\"_col_\"+str(tresh_ratio_gap_col)+\".txt\"\n",
    "    handle_msa_fasta = open(out_name_msa,\"w\")\n",
    "    SeqIO.write(msa_filtered_row_col, handle_msa_fasta,\"fasta\")\n",
    "    handle_msa_fasta.close()\n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Column-wise filtering of MSA is finished\",len(msa_filtered_row_col),len(msa_filtered_row_col[0])) \n",
    "       \n",
    "    #msa_filtered_row_col = MultipleSeqAlignment(msa_filtered_row_col)\n",
    "    return msa_filtered_row_col\n",
    "\n",
    "\n",
    "def draw_tree(msa):\n",
    "    ############## Tree inference  ###################\n",
    "    ##################################################\n",
    "\n",
    "    wrapper_tree=fasttree.Fasttree(msa,datatype=\"PROTEIN\")\n",
    "    wrapper_tree.options.options['-fastest']    \n",
    "    result_tree1 = wrapper_tree()\n",
    "\n",
    "    time_taken_tree = wrapper_tree.elapsed_time \n",
    "    time_taken_tree\n",
    "\n",
    "    result_tree2 = wrapper_tree.result\n",
    "    tree_nwk=str(result_tree2[\"tree\"])\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- \",len(tree_nwk))\n",
    "\n",
    "    out_name_tree=project_folder+\"fastoma_core/_tree.txt\"\n",
    "    file1 = open(out_name_tree,\"w\")\n",
    "    file1.write(tree_nwk)\n",
    "    file1.close() \n",
    "    return tree_nwk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8c023767",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/python3\n",
    "\n",
    "import numpy as np\n",
    "import pyoma.browser.db as db\n",
    "import concurrent.futures\n",
    "import ast\n",
    "import pickle\n",
    "\n",
    "import zoo.wrappers.aligners.mafft as mafft  # mafft should be installed beforehand\n",
    "# import zoo.wrappers.treebuilders.fasttree as fasttree\n",
    "\n",
    "from datetime import datetime\n",
    "from sys import argv\n",
    "from os import listdir\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq, UnknownSeq\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "import matplotlib   #for development \n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "import pyoma.browser.db as db\n",
    "from pyoma.browser.hoghelper import build_hog_to_og_map\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f67b4f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "09:51:21 - OMA data is parsed and its release name is: OmaStandalone; 1.0.x\n",
      "09:51:21 - There are 10 species in the OMA database.\n",
      "OMA data is parsed and its release name is: OmaStandalone; 1.0.x\n",
      "There are 10 species in the OMA database.\n",
      "HOG-OG map is extracted with length of  38795\n",
      "HOG-OG map is written in the file /work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal/mml27/omamer_database/oma_path/OmaServer.h5_hog_og_map.dic\n",
      "09:51:35 - The are 25 species in the proteome folder.\n",
      "09:51:36 - There are  25  in hogmap folder.\n",
      "09:51:36 - The first species CERAT  contains  46067  proteins.\n",
      "09:51:36 - The first protein of first species is  NOMLE00001\n",
      "09:51:44 - For the species GORGO_ , few proteins were ignored by omamer.\n",
      "09:51:44 - before filtering: in hogmap 21821 in proteome 21822\n",
      "09:51:44 - After filtering:  in hogmap 21821 in proteome 21821\n",
      "09:53:58 - For the species MOUSE_ , few proteins were ignored by omamer.\n",
      "09:53:58 - before filtering: in hogmap 67255 in proteome 67347\n",
      "09:53:58 - After filtering:  in hogmap 67255 in proteome 67255\n",
      "10:02:21 - For the species HUMAN , few proteins were ignored by omamer.\n",
      "10:02:21 - before filtering: in hogmap 103039 in proteome 103200\n",
      "10:02:21 - After filtering:  in hogmap 103039 in proteome 103039\n",
      "10:02:21 - For the rest of species, all proteins were mapped using OMAmer.\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "\n",
    "    # a global variable\n",
    "    project_folder = \"/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal/mml27/\"\n",
    "    \n",
    "    oma_database_address = project_folder+\"omamer_database/oma_path/OmaServer.h5\"\n",
    "        \n",
    "    (oma_db, list_speices) = parse_oma_db(oma_database_address)\n",
    "\n",
    "    ## if you already  have the hog-og-map, parse it: \n",
    "    # hog_og_map_address=oma_database_address+\"_hog_og_map.dic\"    \n",
    "    # (hog_OG_map)= parse_hog_og_map(hog_og_map_address)\n",
    "\n",
    "    ## if you you don't have the hog-og-map, run this: \n",
    "    hog_OG_map = HOG_OG_map(oma_database_address)\n",
    "    \n",
    "    \n",
    "    \n",
    "    (query_species_names, query_prot_records_species) = parse_proteome(list_speices)   \n",
    "    (query_prot_names_species_mapped, query_hogids_species) = parse_hogmap_omamer(query_species_names) \n",
    "\n",
    "    query_prot_records_species_filtered=  filter_prot_mapped(query_species_names, query_prot_records_species, query_prot_names_species_mapped)\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c46ce43",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6118fc2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0d70ce63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10:18:40 - Extracting proteins mapped only once on a HOG is started\n",
      "10:18:42 - Extracting proteins mapped only once on a HOG is finished\n",
      "10:18:42 - For  25  species, we keep  267054 HOGs.\n",
      "10:18:45 - Needed HOH-OG map  20812 are extracted from the map file.\n"
     ]
    }
   ],
   "source": [
    "    (query_hogids_filtr_species, query_prot_names_filtr_species, query_prot_records_filtr_species) = extract_unique_hog_pure(query_species_names,query_hogids_species, query_prot_names_species_mapped, query_prot_records_species_filtered) \n",
    "\n",
    "    OGs_queries = gather_OG(query_species_names, query_hogids_filtr_species, query_prot_names_filtr_species, query_prot_records_filtr_species)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b9b9fb7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "56d95af9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23\n",
      "10:21:41 - Start filtering OGs of proteme queries with length of  20812 for  25 species.\n",
      "10:21:41 - Finished filetering from  20812 , now 258 OGs are remained.\n",
      "10:21:41 - These species missed due to the threshold of 23 : []\n"
     ]
    }
   ],
   "source": [
    "    keep_og_treshold_species_query =  int(0.95* len(query_species_names))\n",
    "    print(keep_og_treshold_species_query)\n",
    "    OGs_queries_filtr= filter_gathered_OG(OGs_queries, query_species_names,keep_og_treshold_species_query)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "93a699e9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10:21:42 - Combining queries with OG started for 258 OGs.\n",
      "10:21:44 - Number of OMA species per OG is calculated\n",
      "10:21:44 - These OMA species are kept ['CARSF', 'BOVIN', 'MOUSE', 'OTOGA', 'SAIBB', 'GORGO', 'MANLE', 'COLAP', 'PANTR', 'MACNE']\n",
      "10:21:45  - Combining OG 10905  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 6086  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 6367  with length of  6 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 4547  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 18301  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 3748  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 16366  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 13484  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 24987  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 14975  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 16913  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 20466  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 19584  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 15814  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 4220  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 16652  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 12186  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 6311  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 9671  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 11225  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 20731  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 21405  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 24946  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 21294  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 6830  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 7124  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 3865  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 16686  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 10782  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 19498  with length of  7 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 11376  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 7887  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 8949  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 24916  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 8878  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 13874  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 12031  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 17969  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 16687  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 8556  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 19732  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 7329  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 16536  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 22393  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 19126  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 11673  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 8255  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 9870  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 11058  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 5832  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 16131  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 20888  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 19067  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 25775  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 23772  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 10615  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 10143  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 7475  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 13279  with length of  8 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 19348  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 19665  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 15644  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 17720  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 18074  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 20990  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 12115  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 21519  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 16775  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 13511  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 10678  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 15364  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 8579  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 22833  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 11474  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:45  - Combining OG 14339  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 6021  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 11994  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 18226  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 16473  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 9090  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 17456  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:45  - Combining OG 12695  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 18154  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 9681  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 5727  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 10013  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 21575  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 20873  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 20019  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 12278  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 15391  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 11006  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 7442  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 12433  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 8478  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 19373  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 1843  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 4208  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 11147  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 13672  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 11512  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 13893  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 4484  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 16097  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 15339  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 16569  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 16606  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 19394  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 16547  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 14659  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 2496  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 15804  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 11053  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 14621  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 17512  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 22228  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 14907  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 4309  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 12470  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 13253  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 16873  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 21344  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 18033  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 13368  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 9123  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 20294  with length of  6 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 4646  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 4321  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 18741  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 18558  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 9561  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 6711  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 12240  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 17937  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 17262  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 7788  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 17139  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 8133  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 15045  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 15905  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 14796  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 12070  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 14074  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 16597  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 17498  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 12583  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 13354  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 19474  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 10899  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 9464  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 9436  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 15073  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 17931  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 18888  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 4360  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 11431  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 21513  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:46  - Combining OG 13601  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 15846  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 14592  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 4574  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 3671  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 10245  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:46  - Combining OG 14832  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 12539  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 12872  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 7086  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 7849  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 13161  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 12579  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 12474  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 17637  with length of  8 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 24094  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 11261  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 18776  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 1184  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 11846  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 13648  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 20576  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 17423  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 7763  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 9274  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 9242  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 9221  with length of  8 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 9143  with length of  8 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 16770  with length of  8 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 15708  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 11624  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 23182  with length of  8 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 487  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 7444  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 14272  with length of  8 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 4499  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 17392  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 16477  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 16226  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 17636  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 16695  with length of  8 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 5526  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 10619  with length of  8 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 20122  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 12117  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 16676  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 12613  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 10438  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 12165  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 11850  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 13314  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 15355  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 18581  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 15801  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 6001  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 22222  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 21671  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 22049  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 22764  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 23813  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 23549  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 22686  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 1089  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 15960  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 15631  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 1581  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 8840  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 16791  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 11754  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 13919  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 16279  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 13158  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 1709  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 17711  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 10735  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 9526  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 7724  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 17412  with length of  10 \t with a query  25  is just finished.\n",
      "10:21:47  - Combining OG 8441  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 11578  with length of  8 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 7083  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 9553  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 17325  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 24882  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 18865  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 9577  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 18128  with length of  8 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 4992  with length of  7 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 14675  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 17155  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 20642  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 8621  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:47  - Combining OG 10746  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:48  - Combining OG 10157  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:48  - Combining OG 10065  with length of  9 \t with a query  24  is just finished.\n",
      "10:21:48  - Combining OG 19843  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:48  - Combining OG 5993  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:48  - Combining OG 17668  with length of  9 \t with a query  25  is just finished.\n",
      "10:21:48  - Combining OG 9698  with length of  10 \t with a query  24  is just finished.\n",
      "10:21:48  - Combining OG 12250  with length of  9 \t with a query  24  is just finished.\n",
      "\n",
      " 10:21:48 - Combining queries with OG is finished! number of OGs 257\n",
      "\n",
      " 10:21:48 - The variable seqRecords_all_filtr is saved as /work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal/mml27/fastoma_core/_file_combined_OGs_filtr.pkl\n"
     ]
    }
   ],
   "source": [
    "    threshold_least_query_sepecies_in_OG =  keep_og_treshold_species_query# 15\n",
    "    \n",
    "    kept_oma_species_num = 10 # 20\n",
    "    \n",
    "    seqRecords_all_filtr = combine_OG_query_filtr(OGs_queries_filtr, oma_db, threshold_least_query_sepecies_in_OG,kept_oma_species_num)\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8517196f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f018039b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b31ad9f4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10:21:56 - Parallel msa is started for  257  OGs.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/work/FAC/FBM/DBC/cdessim2/default/smajidi1/software/miniconda3/lib/python3.8/subprocess.py:844: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stdout = io.open(c2pread, 'rb', bufsize)\n",
      "/work/FAC/FBM/DBC/cdessim2/default/smajidi1/software/miniconda3/lib/python3.8/subprocess.py:844: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stdout = io.open(c2pread, 'rb', bufsize)\n",
      "/work/FAC/FBM/DBC/cdessim2/default/smajidi1/software/miniconda3/lib/python3.8/subprocess.py:844: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stdout = io.open(c2pread, 'rb', bufsize)\n",
      "/work/FAC/FBM/DBC/cdessim2/default/smajidi1/software/miniconda3/lib/python3.8/subprocess.py:849: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stderr = io.open(errread, 'rb', bufsize)\n",
      "/work/FAC/FBM/DBC/cdessim2/default/smajidi1/software/miniconda3/lib/python3.8/subprocess.py:849: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stderr = io.open(errread, 'rb', bufsize)\n",
      "/work/FAC/FBM/DBC/cdessim2/default/smajidi1/software/miniconda3/lib/python3.8/subprocess.py:849: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stderr = io.open(errread, 'rb', bufsize)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10:21:57 - MSA for an OG is just finished:  1.2294516563415527\n",
      "10:21:58 - MSA for an OG is just finished:  1.7980847358703613\n",
      "10:21:58 - MSA for an OG is just finished:  0.4918327331542969\n",
      "10:21:58 - MSA for an OG is just finished:  2.3141233921051025\n",
      "10:21:59 - MSA for an OG is just finished:  0.60951828956604\n",
      "10:22:00 - MSA for an OG is just finished:  0.9834351539611816\n",
      "10:22:00 - MSA for an OG is just finished:  0.24334955215454102\n",
      "10:22:01 - MSA for an OG is just finished:  3.438464641571045\n",
      "10:22:01 - MSA for an OG is just finished:  0.7927708625793457\n",
      "10:22:01 - MSA for an OG is just finished:  0.5951099395751953\n",
      "10:22:01 - MSA for an OG is just finished:  0.3484511375427246\n",
      "10:22:02 - MSA for an OG is just finished:  0.47565579414367676\n",
      "10:22:02 - MSA for an OG is just finished:  0.6050050258636475\n",
      "10:22:03 - MSA for an OG is just finished:  0.4945805072784424\n",
      "10:22:03 - MSA for an OG is just finished:  4.499526500701904\n",
      "10:22:04 - MSA for an OG is just finished:  1.0422353744506836\n",
      "10:22:05 - MSA for an OG is just finished:  1.2468023300170898\n",
      "10:22:05 - MSA for an OG is just finished:  3.383638381958008\n",
      "10:22:05 - MSA for an OG is just finished:  2.3436663150787354\n",
      "10:22:06 - MSA for an OG is just finished:  0.3447272777557373\n",
      "10:22:06 - MSA for an OG is just finished:  0.34728360176086426\n",
      "10:22:06 - MSA for an OG is just finished:  0.9808797836303711\n",
      "10:22:06 - MSA for an OG is just finished:  0.254840612411499\n",
      "10:22:06 - MSA for an OG is just finished:  0.3535749912261963\n",
      "10:22:08 - MSA for an OG is just finished:  1.8422138690948486\n",
      "10:22:08 - MSA for an OG is just finished:  2.3213043212890625\n",
      "10:22:08 - MSA for an OG is just finished:  0.5435233116149902\n",
      "10:22:09 - MSA for an OG is just finished:  0.38574767112731934\n",
      "10:22:09 - MSA for an OG is just finished:  1.1196165084838867\n",
      "10:22:10 - MSA for an OG is just finished:  3.6025896072387695\n",
      "10:22:10 - MSA for an OG is just finished:  1.0176208019256592\n",
      "10:22:10 - MSA for an OG is just finished:  0.23202228546142578\n",
      "10:22:11 - MSA for an OG is just finished:  1.4337615966796875\n",
      "10:22:11 - MSA for an OG is just finished:  1.840397596359253\n",
      "10:22:11 - MSA for an OG is just finished:  1.5200316905975342\n",
      "10:22:12 - MSA for an OG is just finished:  0.7660789489746094\n",
      "10:22:12 - MSA for an OG is just finished:  0.5312821865081787\n",
      "10:22:12 - MSA for an OG is just finished:  1.0034043788909912\n",
      "10:22:12 - MSA for an OG is just finished:  0.5811524391174316\n",
      "10:22:13 - MSA for an OG is just finished:  0.4251251220703125\n",
      "10:22:13 - MSA for an OG is just finished:  0.5837488174438477\n",
      "10:22:14 - MSA for an OG is just finished:  0.3228030204772949\n",
      "10:22:14 - MSA for an OG is just finished:  1.6459321975708008\n",
      "10:22:14 - MSA for an OG is just finished:  0.43480896949768066\n",
      "10:22:14 - MSA for an OG is just finished:  2.0543885231018066\n",
      "10:22:15 - MSA for an OG is just finished:  1.0772547721862793\n",
      "10:22:16 - MSA for an OG is just finished:  1.767573356628418\n",
      "10:22:16 - MSA for an OG is just finished:  1.1227474212646484\n",
      "10:22:16 - MSA for an OG is just finished:  1.5033376216888428\n",
      "10:22:16 - MSA for an OG is just finished:  0.39580512046813965\n",
      "10:22:16 - MSA for an OG is just finished:  0.5985352993011475\n",
      "10:22:17 - MSA for an OG is just finished:  0.21816396713256836\n",
      "10:22:17 - MSA for an OG is just finished:  0.42176175117492676\n",
      "10:22:17 - MSA for an OG is just finished:  0.25485944747924805\n",
      "10:22:18 - MSA for an OG is just finished:  1.149669885635376\n",
      "10:22:18 - MSA for an OG is just finished:  2.2744839191436768\n",
      "10:22:18 - MSA for an OG is just finished:  1.1980206966400146\n",
      "10:22:19 - MSA for an OG is just finished:  0.42871904373168945\n",
      "10:22:19 - MSA for an OG is just finished:  0.7903487682342529\n",
      "10:22:19 - MSA for an OG is just finished:  0.44226598739624023\n",
      "10:22:19 - MSA for an OG is just finished:  0.6697015762329102\n",
      "10:22:20 - MSA for an OG is just finished:  0.532282829284668\n",
      "10:22:20 - MSA for an OG is just finished:  0.37588953971862793\n",
      "10:22:20 - MSA for an OG is just finished:  0.5160012245178223\n",
      "10:22:20 - MSA for an OG is just finished:  2.0758068561553955\n",
      "10:22:20 - MSA for an OG is just finished:  0.40847349166870117\n",
      "10:22:21 - MSA for an OG is just finished:  0.6321301460266113\n",
      "10:22:21 - MSA for an OG is just finished:  1.098395586013794\n",
      "10:22:21 - MSA for an OG is just finished:  0.9349329471588135\n",
      "10:22:22 - MSA for an OG is just finished:  0.6255688667297363\n",
      "10:22:22 - MSA for an OG is just finished:  1.2618305683135986\n",
      "10:22:22 - MSA for an OG is just finished:  0.29532790184020996\n",
      "10:22:23 - MSA for an OG is just finished:  0.7475020885467529\n",
      "10:22:23 - MSA for an OG is just finished:  1.0338878631591797\n",
      "10:22:23 - MSA for an OG is just finished:  1.9370646476745605\n",
      "10:22:24 - MSA for an OG is just finished:  0.48298168182373047\n",
      "10:22:24 - MSA for an OG is just finished:  1.0822792053222656\n",
      "10:22:24 - MSA for an OG is just finished:  0.538994312286377\n",
      "10:22:25 - MSA for an OG is just finished:  0.4934225082397461\n",
      "10:22:25 - MSA for an OG is just finished:  2.128861665725708\n",
      "10:22:25 - MSA for an OG is just finished:  0.45726847648620605\n",
      "10:22:26 - MSA for an OG is just finished:  1.4843761920928955\n",
      "10:22:26 - MSA for an OG is just finished:  0.8918752670288086\n",
      "10:22:27 - MSA for an OG is just finished:  1.41241455078125\n",
      "10:22:27 - MSA for an OG is just finished:  1.2738111019134521\n",
      "10:22:27 - MSA for an OG is just finished:  0.3048055171966553\n",
      "10:22:27 - MSA for an OG is just finished:  0.34473752975463867\n",
      "10:22:28 - MSA for an OG is just finished:  0.4811728000640869\n",
      "10:22:28 - MSA for an OG is just finished:  0.6512100696563721\n",
      "10:22:28 - MSA for an OG is just finished:  2.725955009460449\n",
      "10:22:28 - MSA for an OG is just finished:  1.07204008102417\n",
      "10:22:29 - MSA for an OG is just finished:  1.0068488121032715\n",
      "10:22:30 - MSA for an OG is just finished:  1.3856573104858398\n",
      "10:22:30 - MSA for an OG is just finished:  0.43793702125549316\n",
      "10:22:30 - MSA for an OG is just finished:  1.9606099128723145\n",
      "10:22:31 - MSA for an OG is just finished:  1.7787082195281982\n",
      "10:22:33 - MSA for an OG is just finished:  1.413313388824463\n",
      "10:22:33 - MSA for an OG is just finished:  0.7980499267578125\n",
      "10:22:34 - MSA for an OG is just finished:  3.3824455738067627\n",
      "10:22:34 - MSA for an OG is just finished:  0.9491477012634277\n",
      "10:22:35 - MSA for an OG is just finished:  0.9204707145690918\n",
      "10:22:35 - MSA for an OG is just finished:  0.5643007755279541\n",
      "10:22:36 - MSA for an OG is just finished:  0.639472246170044\n",
      "10:22:36 - MSA for an OG is just finished:  0.6380364894866943\n",
      "10:22:37 - MSA for an OG is just finished:  0.5893762111663818\n",
      "10:22:37 - MSA for an OG is just finished:  0.4416205883026123\n",
      "10:22:38 - MSA for an OG is just finished:  0.5416812896728516\n",
      "10:22:38 - MSA for an OG is just finished:  3.850133180618286\n",
      "10:22:39 - MSA for an OG is just finished:  0.7047929763793945\n",
      "10:22:39 - MSA for an OG is just finished:  8.852519273757935\n",
      "10:22:39 - MSA for an OG is just finished:  0.5964159965515137\n",
      "10:22:40 - MSA for an OG is just finished:  0.7249143123626709\n",
      "10:22:40 - MSA for an OG is just finished:  1.2222504615783691\n",
      "10:22:40 - MSA for an OG is just finished:  0.28941965103149414\n",
      "10:22:41 - MSA for an OG is just finished:  0.5372662544250488\n",
      "10:22:41 - MSA for an OG is just finished:  0.730907678604126\n",
      "10:22:42 - MSA for an OG is just finished:  1.1652791500091553\n",
      "10:22:43 - MSA for an OG is just finished:  0.8024554252624512\n",
      "10:22:44 - MSA for an OG is just finished:  0.597337007522583\n",
      "10:22:44 - MSA for an OG is just finished:  5.7172369956970215\n",
      "10:22:44 - MSA for an OG is just finished:  0.35635876655578613\n",
      "10:22:44 - MSA for an OG is just finished:  3.7795634269714355\n",
      "10:22:45 - MSA for an OG is just finished:  0.5212512016296387\n",
      "10:22:45 - MSA for an OG is just finished:  0.43273282051086426\n",
      "10:22:45 - MSA for an OG is just finished:  0.8879272937774658\n",
      "10:22:46 - MSA for an OG is just finished:  1.4106323719024658\n",
      "10:22:46 - MSA for an OG is just finished:  0.4886150360107422\n",
      "10:22:47 - MSA for an OG is just finished:  0.4523913860321045\n",
      "10:22:48 - MSA for an OG is just finished:  1.4510676860809326\n",
      "10:22:48 - MSA for an OG is just finished:  3.4625003337860107\n",
      "10:22:49 - MSA for an OG is just finished:  3.5807700157165527\n",
      "10:22:49 - MSA for an OG is just finished:  0.5812427997589111\n",
      "10:22:49 - MSA for an OG is just finished:  0.9641838073730469\n",
      "10:22:50 - MSA for an OG is just finished:  0.5770273208618164\n",
      "10:22:50 - MSA for an OG is just finished:  1.8829951286315918\n",
      "10:22:50 - MSA for an OG is just finished:  0.5629439353942871\n",
      "10:22:51 - MSA for an OG is just finished:  0.7920379638671875\n",
      "10:22:51 - MSA for an OG is just finished:  2.066877603530884\n",
      "10:22:52 - MSA for an OG is just finished:  1.7434608936309814\n",
      "10:22:52 - MSA for an OG is just finished:  0.806833028793335\n",
      "10:22:52 - MSA for an OG is just finished:  0.6860082149505615\n",
      "10:22:53 - MSA for an OG is just finished:  0.7878572940826416\n",
      "10:22:5310:22:53  - MSA for an OG is just finished: - MSA for an OG is just finished:   0.90378379821777340.617633581161499\n",
      "\n",
      "10:22:53 - MSA for an OG is just finished:  0.566448450088501\n",
      "10:22:54 - MSA for an OG is just finished:  0.8464944362640381\n",
      "10:22:54 - MSA for an OG is just finished:  0.4631819725036621\n",
      "10:22:54 - MSA for an OG is just finished:  1.0160701274871826\n",
      "10:22:55 - MSA for an OG is just finished:  1.1522130966186523\n",
      "10:22:55 - MSA for an OG is just finished:  1.3624498844146729\n",
      "10:22:55 - MSA for an OG is just finished:  1.5098397731781006\n",
      "10:22:55 - MSA for an OG is just finished:  0.6479370594024658\n",
      "10:22:56 - MSA for an OG is just finished:  0.5326423645019531\n",
      "10:22:56 - MSA for an OG is just finished:  0.44327425956726074\n",
      "10:22:56 - MSA for an OG is just finished:  0.37943601608276367\n",
      "10:22:57 - MSA for an OG is just finished:  1.1159241199493408\n",
      "10:22:57 - MSA for an OG is just finished:  1.0145187377929688\n",
      "10:22:57 - MSA for an OG is just finished:  0.5715928077697754\n",
      "10:22:58 - MSA for an OG is just finished:  0.6919283866882324\n",
      "10:22:59 - MSA for an OG is just finished:  3.869023561477661\n",
      "10:23:01 - MSA for an OG is just finished:  1.2644662857055664\n",
      "10:23:01 - MSA for an OG is just finished:  3.6914103031158447\n",
      "10:23:01 - MSA for an OG is just finished:  0.698474645614624\n",
      "10:23:02 - MSA for an OG is just finished:  0.9016914367675781\n",
      "10:23:02 - MSA for an OG is just finished:  4.189495801925659\n",
      "10:23:02 - MSA for an OG is just finished:  0.8771517276763916\n",
      "10:23:03 - MSA for an OG is just finished:  0.9058713912963867\n",
      "10:23:04 - MSA for an OG is just finished:  0.9386701583862305\n",
      "10:23:04 - MSA for an OG is just finished:  2.100908041000366\n",
      "10:23:04 - MSA for an OG is just finished:  2.3602023124694824\n",
      "10:23:05 - MSA for an OG is just finished:  0.29396867752075195\n",
      "10:23:05 - MSA for an OG is just finished:  0.6129043102264404\n",
      "10:23:05 - MSA for an OG is just finished:  1.0281617641448975\n",
      "10:23:05 - MSA for an OG is just finished:  0.54343581199646\n",
      "10:23:06 - MSA for an OG is just finished:  1.0184893608093262\n",
      "10:23:06 - MSA for an OG is just finished:  1.001105546951294\n",
      "10:23:06 - MSA for an OG is just finished:  0.7980446815490723\n",
      "10:23:07 - MSA for an OG is just finished:  0.4407467842102051\n",
      "10:23:07 - MSA for an OG is just finished:  0.5146470069885254\n",
      "10:23:08 - MSA for an OG is just finished:  1.4921410083770752\n",
      "10:23:09 - MSA for an OG is just finished:  2.1093013286590576\n",
      "10:23:10 - MSA for an OG is just finished:  1.7495968341827393\n",
      "10:23:10 - MSA for an OG is just finished:  1.497880220413208\n",
      "10:23:11 - MSA for an OG is just finished:  0.5561723709106445\n",
      "10:23:12 - MSA for an OG is just finished:  0.603731632232666\n",
      "10:23:12 - MSA for an OG is just finished:  1.3536860942840576\n",
      "10:23:12 - MSA for an OG is just finished:  0.32801270484924316\n",
      "10:23:13 - MSA for an OG is just finished:  1.1136460304260254\n",
      "10:23:15 - MSA for an OG is just finished:  1.9249343872070312\n",
      "10:23:15 - MSA for an OG is just finished:  10.140770673751831\n",
      "10:23:15 - MSA for an OG is just finished:  0.8054988384246826\n",
      "10:23:16 - MSA for an OG is just finished:  0.5918207168579102\n",
      "10:23:17 - MSA for an OG is just finished:  0.5781214237213135\n",
      "10:23:17 - MSA for an OG is just finished:  0.6347596645355225\n",
      "10:23:18 - MSA for an OG is just finished:  0.5225882530212402\n",
      "10:23:18 - MSA for an OG is just finished:  3.105639696121216\n",
      "10:23:18 - MSA for an OG is just finished:  0.6052100658416748\n",
      "10:23:20 - MSA for an OG is just finished:  1.5091123580932617\n",
      "10:23:20 - MSA for an OG is just finished:  0.4197559356689453\n",
      "10:23:21 - MSA for an OG is just finished:  2.561396360397339\n",
      "10:23:21 - MSA for an OG is just finished:  1.021064043045044\n",
      "10:23:21 - MSA for an OG is just finished:  0.5959675312042236\n",
      "10:23:23 - MSA for an OG is just finished:  1.2379326820373535\n",
      "10:23:23 - MSA for an OG is just finished:  1.2459077835083008\n",
      "10:23:24 - MSA for an OG is just finished:  1.1055388450622559\n",
      "10:23:24 - MSA for an OG is just finished:  1.0172820091247559\n",
      "10:23:24 - MSA for an OG is just finished:  0.6555497646331787\n",
      "10:23:25 - MSA for an OG is just finished:  0.8614346981048584\n",
      "10:23:25 - MSA for an OG is just finished:  0.4975154399871826\n",
      "10:23:25 - MSA for an OG is just finished:  0.5932250022888184\n",
      "10:23:25 - MSA for an OG is just finished:  0.34734010696411133\n",
      "10:23:26 - MSA for an OG is just finished:  0.35604190826416016\n",
      "10:23:26 - MSA for an OG is just finished:  0.35964059829711914\n",
      "10:23:27 - MSA for an OG is just finished:  0.29001903533935547\n",
      "10:23:27 - MSA for an OG is just finished:  0.3013784885406494\n",
      "10:23:27 - MSA for an OG is just finished:  0.304121732711792\n",
      "10:23:27 - MSA for an OG is just finished:  0.29518914222717285\n",
      "10:23:27 - MSA for an OG is just finished:  2.58498477935791\n",
      "10:23:28 - MSA for an OG is just finished:  0.6286520957946777\n",
      "10:23:29 - MSA for an OG is just finished:  0.5818028450012207\n",
      "10:23:36 - MSA for an OG is just finished:  24.12398409843445\n",
      "10:23:37 - MSA for an OG is just finished:  8.329907178878784\n",
      "10:23:38 - MSA for an OG is just finished:  1.6591899394989014\n",
      "10:23:38 - MSA for an OG is just finished:  0.681647777557373\n",
      "10:23:39 - MSA for an OG is just finished:  0.8145596981048584\n",
      "10:23:39 - MSA for an OG is just finished:  1.032691478729248\n",
      "10:23:39 - MSA for an OG is just finished:  0.6889691352844238\n",
      "10:23:40 - MSA for an OG is just finished:  0.908552885055542\n",
      "10:23:40 - MSA for an OG is just finished:  12.27117919921875\n",
      "10:23:40 - MSA for an OG is just finished:  0.46315693855285645\n",
      "10:23:41 - MSA for an OG is just finished:  1.2918925285339355\n",
      "10:23:42 - MSA for an OG is just finished:  1.5549259185791016\n",
      "10:23:42 - MSA for an OG is just finished:  0.5922346115112305\n",
      "10:23:43 - MSA for an OG is just finished:  1.8771121501922607\n",
      "10:23:44 - MSA for an OG is just finished:  1.1608095169067383\n",
      "10:23:44 - MSA for an OG is just finished:  1.8248867988586426\n",
      "10:23:46 - MSA for an OG is just finished:  1.4898691177368164\n",
      "10:23:46 - MSA for an OG is just finished:  1.8782541751861572\n",
      "10:23:46 - MSA for an OG is just finished:  0.23003888130187988\n",
      "10:23:46 - MSA for an OG is just finished:  0.572744607925415\n",
      "10:23:47 - MSA for an OG is just finished:  0.4429342746734619\n",
      "10:23:47 - MSA for an OG is just finished:  0.4695141315460205\n",
      "10:23:48 - MSA for an OG is just finished:  1.4156386852264404\n",
      "10:23:48 - MSA for an OG is just finished:  8.464609146118164\n",
      "10:23:48 - MSA for an OG is just finished:  0.6147618293762207\n",
      "10:23:49 - MSA for an OG is just finished:  0.9054162502288818\n",
      "10:23:49 - MSA for an OG is just finished:  0.3504068851470947\n",
      "10:23:50 - MSA for an OG is just finished:  2.4844858646392822\n",
      "10:23:50 - MSA for an OG is just finished:  1.6571242809295654\n",
      "10:23:50 - MSA for an OG is just finished:  1.6045098304748535\n",
      "10:23:51 - MSA for an OG is just finished:  0.44728970527648926\n",
      "10:23:51 - MSA for an OG is just finished:  1.4969773292541504\n",
      "10:23:51 - MSA for an OG is just finished:  1.2936904430389404\n",
      "10:23:52 - MSA for an OG is just finished:  0.47322916984558105\n",
      "10:23:53 - MSA for an OG is just finished:  1.0240154266357422\n",
      "10:23:53 - MSA for an OG is just finished:  1.5449130535125732\n",
      "10:23:53 - MSA for an OG is just finished:  2.5785889625549316\n",
      "All input species are included in the MSA.\n",
      "10:23:53 - Filtering MSA started.\n",
      "10:23:54 - The histogram of density ogs is saved.\n",
      "10:23:54 - Filtering MSA finished, keeping 100 out of 257\n",
      "10:23:54 - MSA concatenation started\n",
      "10:23:54 - Alignments len 100\n",
      "10:23:54 - ids:  35 3353\n",
      "10:23:54 - MSA concatenation finished\n",
      "10:23:54 - MSA concatenation has been written in the file /work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal/mml27/fastoma_core/_100_msa_concatanated.txt with length 35 37539\n"
     ]
    }
   ],
   "source": [
    " \n",
    "    number_max_workers = 3 # round(int(argv[3])*1.3)\n",
    "    result_mafft_all_species = run_msa_OG_parallel(seqRecords_all_filtr,number_max_workers)\n",
    "\n",
    "    \n",
    "    all_id=[]\n",
    "    for a in result_mafft_all_species:\n",
    "        for i in a:\n",
    "            all_id.append(i.id)\n",
    "    len(all_id),len(set(all_id))\n",
    "\n",
    "    list_removed_species=[]\n",
    "    for i in query_species_names:\n",
    "        if i not in set(all_id):\n",
    "            list_removed_species.append(i)\n",
    "\n",
    "    if len(list_removed_species):\n",
    "        print(\"These species are removed from the input proteome:\",list_removed_species)\n",
    "    else:\n",
    "        print(\"All input species are included in the MSA.\")\n",
    "\n",
    "    list_added_species=[]\n",
    "    for i in set(all_id):\n",
    "        if i not in query_species_names:\n",
    "            list_added_species.append(i)        \n",
    "    if len(list_removed_species):\n",
    "        print(\"These species are added from OMA database:\",list_added_species)\n",
    "\n",
    "\n",
    "        \n",
    "        \n",
    "    ogs_keep_number = 100\n",
    "    result_mafft_all_species_filtr = filter_ogs(result_mafft_all_species,ogs_keep_number)\n",
    "    msa= concatante_alignments(result_mafft_all_species_filtr, ogs_keep_number)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ec23946",
   "metadata": {},
   "source": [
    "## Load pkl file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92625d07",
   "metadata": {},
   "outputs": [],
   "source": [
    "# a=\"/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal/mml_v4/omamer_v4_search/fastoma_out/\"\n",
    "# a+= \"_file_msas.pkl\"\n",
    "# open_file = open(a, \"rb\") #project_folder+\"_file_msas2.pkl\"\n",
    "# result_mafft_all_species = pickle.load(open_file)\n",
    "# open_file.close()\n",
    "# print(\"seq read is loaded\", len(result_mafft_all_species))\n",
    "\n",
    "# result_mafft_all_species[0][:]\n",
    "# ogs_keep_number = 82\n",
    "# result_mafft_all_species_filtr = filter_ogs(result_mafft_all_species,ogs_keep_number)\n",
    "# msa= concatante_alignments(result_mafft_all_species_filtr, ogs_keep_number)\n",
    "\n",
    "\n",
    "# for record in msa:\n",
    "#     if record.id ==\"S0003\":\n",
    "#         a= record.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93df8aa2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e795a95c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84925a29",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36432298",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b4a2790",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "187e4932",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# reading pkl file \n",
    "# open_file = open(project_folder+\"_file_msas2.pkl\", \"rb\")\n",
    "# result_mafft_all_species = pickle.load(open_file)\n",
    "# open_file.close()\n",
    "# print(\"seq read is loaded\", len(seqRecords_all))\n",
    "\n",
    "#print(current_time, \"- Row-wise filtering of MSA is started.\")        \n",
    "#tresh_ratio_gap_row = 0.3\n",
    "#msa_filtered_row = msa_filter_row(msa,tresh_ratio_gap_row,query_species_names,ogs_keep_number)\n",
    "#print(current_time, \"- Column-wise filtering of MSA is started.\") \n",
    "#tresh_ratio_gap_col = 0.3\n",
    "#msa_filtered_row_col=  msa_filter_col(msa_filtered_row, tresh_ratio_gap_col,tresh_ratio_gap_row,ogs_keep_number)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e178fcdc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a06c87c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c69c9ff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a99382e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#oma_database_address = \"/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal/mml_v4/omamer_v4_database/oma_path_/OmaServer.h5\"\n",
    "#oma_db = db.Database(oma_database_address+\"\")\n",
    "\n",
    "#oma_db.tax.as_dict()\n",
    "#oma_db.tax.genomes.values()\n",
    "#oma_db.tax.newick()\n",
    "#import pyoma.browser.models as models\n",
    "# models.Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dc3ec72",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
