{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5004f57c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bfe573f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/python3\n",
    "\n",
    "\n",
    "# IMPORTNAT  this code accept only .fa file as proteome\n",
    "\n",
    "def parse_oma_db(oma_database_address):\n",
    "    \n",
    "    ############### Parsing OMA db ####################\n",
    "    ###################################################\n",
    "\n",
    "    oma_db = db.Database(oma_database_address)\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- OMA data is parsed and its release name is:\", oma_db.get_release_name())\n",
    "    list_oma_speices = [z.uniprot_species_code for z in oma_db.tax.genomes.values()] \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- There are\",len(list_oma_speices),\"species in the OMA database.\")\n",
    "    \n",
    "    return (oma_db, list_oma_speices)\n",
    "\n",
    "\n",
    "def parse_hog_og_map(hog_og_map_address):\n",
    "    \n",
    "    ############### Parsing hog map python dic #############\n",
    "    #########################################################\n",
    "\n",
    "    oma_db = db.Database(oma_database_address)\n",
    "\n",
    "    file = open(hog_og_map_address, \"r\")\n",
    "    contents = file.read()\n",
    "    hog_OG_map = ast.literal_eval(contents)\n",
    "    file.close()\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- The hog-og map is read from file with the length of \", len(hog_OG_map))\n",
    "       \n",
    "    return (hog_OG_map)\n",
    "\n",
    "\n",
    "\n",
    "def HOG_OG_map(oma_database_address):\n",
    "\n",
    "    oma_db = db.Database(oma_database_address)\n",
    "    print(\"OMA data is parsed and its release name is:\", oma_db.get_release_name())\n",
    "    list_hogog_speices = [z.uniprot_species_code for z in oma_db.tax.genomes.values()] \n",
    "    print(\"There are\",len(list_hogog_speices),\"species in the HOG-OG OMA database.\")\n",
    "\n",
    "    hog_OG_map = {hog_id.decode(\"utf-8\") : og_count_list[0][0] for hog_id, og_count_list in build_hog_to_og_map(oma_db)}\n",
    "    print(\"HOG-OG map is extracted with length of \", len(hog_OG_map))\n",
    "\n",
    "\n",
    "    hog_OG_map_address = oma_database_address+\"_hog_og_map.dic\"\n",
    "    with open(hog_OG_map_address, 'w') as file_map:\n",
    "        print(hog_OG_map, file = file_map)\n",
    "\n",
    "    file_map.close()\n",
    "    print(\"HOG-OG map is written in the file\",hog_OG_map_address)\n",
    "    \n",
    "    return (hog_OG_map)\n",
    "\n",
    "\n",
    "def parse_proteome(list_oma_speices):\n",
    "    \n",
    "    ############### Parsing query proteome of species #######\n",
    "    #########################################################\n",
    "\n",
    "    \n",
    "    # IMPORTNAT  this code accept only .fa file as proteome\n",
    "    project_files = listdir(project_folder+\"/omamer_search/proteome/\")\n",
    "\n",
    "    query_species_names = []\n",
    "    for file in project_files:\n",
    "        if file.split(\".\")[-1] == \"fa\":\n",
    "            file_name_split = file.split(\".\")[:-1]\n",
    "            query_species_names.append('.'.join(file_name_split))\n",
    "        if file.split(\".\")[-1] == \"fasta\":\n",
    "            file_name_split = file.split(\".\")[:-1]\n",
    "            query_species_names.append('.'.join(file_name_split))\n",
    "\n",
    "    # we may assert existence of query_species_name+\".fa/hogmap\"\n",
    "    prots_record_allspecies = [ ]\n",
    "    for query_species_name in query_species_names:\n",
    "        prot_address = project_folder +\"omamer_search/proteome/\"+ query_species_name + \".fa\" \n",
    "        \n",
    "        prots_record = list(SeqIO.parse(prot_address, \"fasta\")) \n",
    "        prots_record_allspecies.append(prots_record)\n",
    "\n",
    "    query_species_num = len(query_species_names)    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- The are\",str(query_species_num),\"species in the proteome folder.\")\n",
    "\n",
    "    # for development\n",
    "    for species_i in range(query_species_num):\n",
    "        len_prot_record_i = len( prots_record_allspecies[species_i] )\n",
    "        species_name_i = query_species_names[species_i]\n",
    "        #print(species_name_i,len_prot_record_i)\n",
    "        if species_name_i in list_oma_speices: \n",
    "            current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "            print(current_time,\"- the species\",species_name_i,\" already exists in the oma database, remove them first\")\n",
    "            exit()\n",
    "\n",
    "      \n",
    "    # The proteins are parsed using  Bio.SeqIO.parse\n",
    "    # the first part of the header line before space \n",
    "    # >tr|A0A2I3FYY2|A0A2I3FYY2_NOMLE Uncharacterized protein OS=Nomascus leucogenys OX=61853 GN=CLPTM1L PE=3 SV=1\n",
    "    # will be \">tr|A0A2I3FYY2|A0A2I3FYY2_NOMLE\"\n",
    "    # [i.id for i in prots_record_allspecies[0] if len(i.id)!=30 and len(i.id)!=22 ] #'sp|O47892|CYB_NOMLE',\n",
    "    \n",
    "\n",
    "    \n",
    "    return (query_species_names, prots_record_allspecies)\n",
    "\n",
    "def parse_splice_file(query_species_names):\n",
    "\n",
    "    protnames_splices_prots_allspecies = []\n",
    "    species_name_noSplice = []\n",
    "    for query_species_name in query_species_names:\n",
    "        splice_address = project_folder +\"omamer_search/proteome/\"+ query_species_name + \".splice\" \n",
    "\n",
    "        protnames_splices_prots = []\n",
    "\n",
    "        try:\n",
    "            splice_file = open(splice_address,'r');    \n",
    "            for line in splice_file:\n",
    "                line_strip = line.strip()\n",
    "                protnames_splices_prot = line_strip.split(\";\") \n",
    "                protnames_splices_prots.append(protnames_splices_prot)\n",
    "        except: \n",
    "            species_name_noSplice.append(query_species_name)\n",
    "\n",
    "        protnames_splices_prots_allspecies.append(protnames_splices_prots)\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- No splicing file found for these species: \", species_name_noSplice)\n",
    "\n",
    "    return protnames_splices_prots_allspecies\n",
    "\n",
    "\n",
    "\n",
    "def parse_hogmap_omamer(query_species_names):\n",
    "\n",
    "    ################### Parsing omamer's output  ########\n",
    "    #####################################################\n",
    "    \n",
    "    prots_hogmap_name_allspecies = []\n",
    "    prots_hogmap_hogid_allspecies = []\n",
    "    prots_hogmap_subfscore_allspecies = []\n",
    "    prots_hogmap_seqlen_allspecies = []\n",
    "    prots_hogmap_subfmedseqlen_allspecies = []\n",
    "\n",
    "    for query_species_name in query_species_names:\n",
    "        omamer_output_address = project_folder + \"omamer_search/hogmap/\"+ query_species_name + \".hogmap\"     \n",
    "        omamer_output_file = open(omamer_output_address,'r');\n",
    "        prots_hogmap_name = []\n",
    "        prots_hogmap_hogid = []\n",
    "        prots_hogmap_subfscore = []\n",
    "        prots_hogmap_seqlen = []\n",
    "        prots_hogmap_subfmedseqlen = []\n",
    "        \n",
    "        for line in omamer_output_file:\n",
    "            line_strip=line.strip()\n",
    "            if not line_strip.startswith('qs'):\n",
    "                line_split= line_strip.split(\"\\t\")    \n",
    "                #if line_split[1]!='na':\n",
    "                prots_hogmap_name.append(line_split[0])\n",
    "                prots_hogmap_hogid.append(line_split[1])\n",
    "                prots_hogmap_subfscore.append(line_split[4]) # subfamily\n",
    "                prots_hogmap_seqlen.append(line_split[5])\n",
    "                prots_hogmap_subfmedseqlen.append(line_split[6])\n",
    "                \n",
    "        prots_hogmap_name_allspecies.append(prots_hogmap_name)\n",
    "        prots_hogmap_hogid_allspecies.append(prots_hogmap_hogid)\n",
    "        prots_hogmap_subfscore_allspecies.append(prots_hogmap_subfscore)\n",
    "        prots_hogmap_seqlen_allspecies.append(prots_hogmap_seqlen)\n",
    "        prots_hogmap_subfmedseqlen_allspecies.append(prots_hogmap_subfmedseqlen)\n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- There are \",len(prots_hogmap_name_allspecies),\" species in the hogmap folder.\")\n",
    "    print(current_time,\"- The first species\",query_species_names[0],\" contains \",len(prots_hogmap_hogid_allspecies[0]),\" proteins.\")\n",
    "    print(current_time,\"- The first protein of first species is \", prots_hogmap_name_allspecies[0][0])\n",
    "\n",
    "    hogmap_allspecies = (prots_hogmap_name_allspecies, prots_hogmap_hogid_allspecies, prots_hogmap_subfscore_allspecies, prots_hogmap_seqlen_allspecies, prots_hogmap_subfmedseqlen_allspecies)\n",
    "    return  hogmap_allspecies\n",
    "    \n",
    "    \n",
    "\n",
    "def filter_prot_mapped(query_species_names, query_prot_records_species,query_prot_names_species_mapped):\n",
    "    # omamer remove very small proteins, \n",
    "    # so  we lose track of order comparing hogmap and fasta file\n",
    "    # the goal here is to remove those from seq record (of the fasta file)\n",
    "    query_prot_records_species_filtered=[]\n",
    "    for species_idx in range(len(query_species_names)):    \n",
    "        # from fasta file\n",
    "        query_species_name=query_species_names[species_idx]\n",
    "        query_prot_records_species_i = query_prot_records_species[species_idx]\n",
    "        query_prot_ids_records = [record.id for record in query_prot_records_species_i]\n",
    "\n",
    "        # from hogmap file\n",
    "        # without proteins that are not mapped on any hogs\n",
    "        query_prot_names_species_i = query_prot_names_species_mapped[species_idx]\n",
    "\n",
    "        if len(query_prot_names_species_i) != len(query_prot_records_species_i):\n",
    "\n",
    "            query_prot_records_filterd=[]\n",
    "            for query_prot_name in query_prot_names_species_i:\n",
    "                if query_prot_name in query_prot_ids_records:\n",
    "                    prot_record_idx = query_prot_ids_records.index(query_prot_name)\n",
    "                    prot_record = query_prot_records_species_i[prot_record_idx]\n",
    "                    query_prot_records_filterd.append(prot_record)\n",
    "                else:\n",
    "                    current_time = datetime.now().strftime(\"%H:%M:%S\")                    \n",
    "                    print(current_time,\"- Error\",query_species_name, query_prot_name)\n",
    "\n",
    "            current_time = datetime.now().strftime(\"%H:%M:%S\")        \n",
    "            print(current_time,\"- For the species\", query_species_name, \", few proteins were ignored by omamer.\")\n",
    "            print(current_time,\"- before filtering: in hogmap\", len(query_prot_names_species_i), \"in proteome\", len(query_prot_records_species_i))\n",
    "            print(current_time,\"- After filtering:  in hogmap\", len(query_prot_names_species_i), \"in proteome\", len(query_prot_records_filterd))            \n",
    "            \n",
    "\n",
    "        else:\n",
    "            query_prot_records_filterd = query_prot_records_species_i\n",
    "\n",
    "        query_prot_records_species_filtered.append(query_prot_records_filterd)\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")        \n",
    "    print(current_time,\"- For the rest of species, all proteins were mapped using OMAmer.\")\n",
    "\n",
    "    return query_prot_records_species_filtered\n",
    "\n",
    "\n",
    "\n",
    "def extract_unique_hog_pure(query_species_names,query_hogids_species, query_prot_names_species_mapped,query_prot_records_species):\n",
    "    ###### Extracting unique HOG list and corresponding query proteins ########\n",
    "    ###########################################################################\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- Extracting proteins mapped only once on a HOG is started\")\n",
    "\n",
    "    query_hogids_filtr_species = []\n",
    "    query_prot_names_filtr_species = []\n",
    "    query_prot_records_filtr_species = []\n",
    "\n",
    "    repeated_hogid_num = 0\n",
    "    \n",
    "    query_species_num = len(query_species_names) \n",
    "    \n",
    "    dic_hogs_list=[]  # a list of dictinaries \n",
    "\n",
    "\n",
    "    for species_i in range(query_species_num):\n",
    "\n",
    "        query_hogids =  query_hogids_species[species_i]\n",
    "        query_prot_names = query_prot_names_species_mapped[species_i]\n",
    "        query_prot_records  = query_prot_records_species[species_i]\n",
    "        \n",
    "        dic_hogs = {}\n",
    "        for prot_i in range(len(query_hogids)):\n",
    "            query_hogid      = query_hogids[prot_i]\n",
    "            query_prot_name  = query_prot_names[prot_i]\n",
    "            query_prot_record= query_prot_records[prot_i]\n",
    "            if query_hogid!= 'na':\n",
    "                if  query_hogid  not in  dic_hogs:\n",
    "                    dic_hogs[query_hogid]=[(query_prot_name,query_prot_record)]\n",
    "                else:\n",
    "                    repeated_hogid_num += 1 \n",
    "                    dic_hogs[query_hogid].append((query_prot_name,query_prot_record))\n",
    "        dic_hogs_list.append(dic_hogs)\n",
    "\n",
    "        \n",
    "    for dic_hogs in dic_hogs_list: # each species\n",
    "        \n",
    "        query_prot_names_filtr = []\n",
    "        query_prot_records_filtr = []\n",
    "\n",
    "        hogid_list = list(dic_hogs.keys())\n",
    "        for hogid in hogid_list:\n",
    "            list_query_prot = dic_hogs[hogid]\n",
    "            if len(list_query_prot)>1:\n",
    "                del dic_hogs[hogid]\n",
    "        #here dic_hogs is updated and  dic_hogs_list  is also updated.\n",
    "        \n",
    "        #print(len(hogid_list),len(dic_hogs))\n",
    "        #for key, value in d.items():\n",
    "        query_hogids_filtr = []\n",
    "        query_prot_names_filtr = []\n",
    "        query_prot_records_filtr = []        \n",
    "        #query_hogids_filtr=list(dic_hogs.keys())\n",
    "        for hogid, tuple_value  in dic_hogs.items():\n",
    "            query_hogids_filtr.append(hogid)\n",
    "            query_prot_names_filtr.append(tuple_value[0][0])\n",
    "            query_prot_records_filtr.append(tuple_value[0][1])\n",
    "        \n",
    "        \n",
    "        \n",
    "\n",
    "        query_hogids_filtr_species.append(query_hogids_filtr)\n",
    "        query_prot_names_filtr_species.append(query_prot_names_filtr)\n",
    "        query_prot_records_filtr_species.append(query_prot_records_filtr)        \n",
    "    \n",
    "    current_time  = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time,\"- Extracting proteins mapped only once on a HOG is finished\")\n",
    "    num_all_hogs=np.sum([len(dic_hogs) for dic_hogs in dic_hogs_list])\n",
    "    print(current_time,\"- For \",len(dic_hogs_list),\" species, we keep \",num_all_hogs,\"HOGs.\")\n",
    "    \n",
    "\n",
    "    return (query_hogids_filtr_species, query_prot_names_filtr_species, query_prot_records_filtr_species )\n",
    "  \n",
    "\n",
    "def gather_OG(query_species_names, query_hogids_filtr_species, query_prot_names_filtr_species, query_prot_records_filtr_species):\n",
    "\n",
    "    ############ Extracting the most frequent OG  ########\n",
    "    #####################################################\n",
    "\n",
    "    #dict (oma_group_nr -> dict(species, [proteins]))\n",
    "    #Og[555] = {homo_erectus: [blabla, blublu], yellow_bird: [P52134], brown_bear: [P2121,B53223]}\n",
    "\n",
    "    OGs_queries = {}\n",
    "\n",
    "    # hog_OG_map = {}\n",
    "\n",
    "    mostFrequent_OG_list_species = []\n",
    "\n",
    "    frq_most_frequent_og_list_all = [] # for development\n",
    "    \n",
    "    query_species_num = len(query_species_names)  \n",
    "    for species_i in  range(query_species_num):\n",
    "\n",
    "        query_species_name = query_species_names[species_i]\n",
    "        #print(\"\\n\",query_species_name)\n",
    "\n",
    "        query_hogids_filtr = query_hogids_filtr_species[species_i]\n",
    "        query_prot_names_filtr = query_prot_names_filtr_species[species_i]\n",
    "        query_prot_records_filtr = query_prot_records_filtr_species[species_i]\n",
    "\n",
    "        mostFrequent_OG_list=[]\n",
    "\n",
    "        num_query_filtr = len(query_hogids_filtr)\n",
    "        for  item_idx in range(num_query_filtr): #\n",
    "\n",
    "            #query_protein = query_prot_names_filtr[item_idx]\n",
    "            seqRecords_query =  query_prot_records_filtr[item_idx]\n",
    "            seqRecords_query_edited = SeqRecord(Seq(str(seqRecords_query.seq)), query_species_name, '', '')\n",
    "            #print(seqRecords_query_edited)\n",
    "\n",
    "            hog_id= query_hogids_filtr[item_idx]\n",
    "\n",
    "            if not hog_id in hog_OG_map:   # Caculitng  most frq OG for the new hog\n",
    "                mostFrequent_OG= -1\n",
    "                hog_OG_map[hog_id]=mostFrequent_OG\n",
    "\n",
    "            else:  # hog_id is in hog_OG_map dic\n",
    "                #print(\"using the hog-og-map\")\n",
    "                mostFrequent_OG = hog_OG_map[hog_id]\n",
    "\n",
    "            if mostFrequent_OG in OGs_queries:\n",
    "                OGs_queries_k = OGs_queries[mostFrequent_OG]\n",
    "\n",
    "                if not query_species_name in OGs_queries_k:\n",
    "                    OGs_queries_k[query_species_name] = seqRecords_query_edited\n",
    "                    OGs_queries[mostFrequent_OG] = OGs_queries_k\n",
    "            else:\n",
    "                OGs_queries[mostFrequent_OG] = {query_species_name: seqRecords_query_edited} # query_protein = query_prot_names_filtr[item_idx]\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Needed HOH-OG map \",len(OGs_queries),\"are extracted from the map file.\") \n",
    "    \n",
    "    return OGs_queries\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "# def filter_gathered_OG(OGs_queries, query_species_names, keep_og_treshold_species_query):\n",
    "\n",
    "    \n",
    "#     current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "#     print(current_time, \"- Start filtering OGs of proteme queries with length of \", len(OGs_queries),\"for \",len(query_species_names),\"species.\") \n",
    "        \n",
    "#     OGs_queries_filtr={}\n",
    "\n",
    "#     query_species_og_kept_set= set()\n",
    "\n",
    "#     query_species_num_OGs_list=[]\n",
    "#     for OG_id, values in OGs_queries.items():\n",
    "#         query_species_og= list(values.keys())\n",
    "\n",
    "#         query_species_num_OGs_list.append(len(query_species_og)) \n",
    "#         if len(query_species_og) > keep_og_treshold_species_query:\n",
    "#             OGs_queries_filtr[OG_id]=values\n",
    "#             query_species_og_kept_set= query_species_og_kept_set.union(set(query_species_og))\n",
    "\n",
    "\n",
    "#     current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "#     print(current_time, \"- Finished filetering from \",len(OGs_queries),\", now\",len(OGs_queries_filtr),\"OGs are remained.\") \n",
    "               \n",
    "\n",
    "\n",
    "#     not_included_species =[]\n",
    "#     for i in set(query_species_names):\n",
    "#         if i not in query_species_og_kept_set:\n",
    "#             not_included_species.append(i)\n",
    "\n",
    "    \n",
    "#     current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "#     print(current_time, \"- These species missed due to the threshold of\",keep_og_treshold_species_query,\":\",not_included_species) \n",
    "               \n",
    "\n",
    "\n",
    "#     #query_species_num_OGs_list\n",
    "#     # for i in not_included_species:\n",
    "#     #     #print(i)\n",
    "#     #     for OG_id, values in OGs_queries.items():\n",
    "#     #         if i in values:\n",
    "#     #             print(values.keys())\n",
    "#     # easiet way keep the first one or make list and keep the best\n",
    "#     # but better to set trehsold not to misss any species \n",
    "    \n",
    "#     return OGs_queries_filtr\n",
    "    \n",
    "\n",
    "\n",
    "def filter_gathered_OG_2(OGs_queries, query_species_names, num_OG_keep_before_msa):\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Start filtering OGs of proteme queries with length of \", len(OGs_queries),\"for \",len(query_species_names),\"species.\") \n",
    "     \n",
    "    \n",
    "    if len(OGs_queries) < num_OG_keep_before_msa:\n",
    "        \n",
    "        OGs_queries_filtr = OGs_queries   \n",
    "        \n",
    "        current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "        print(current_time, \"- Number of OGs is \",len(OGs_queries),\",which is too low, so filtering is not needed.\") \n",
    "    \n",
    "    else:        \n",
    "    \n",
    "        OGs_queries_filtr={}\n",
    "\n",
    "        query_species_og_kept_set= set()\n",
    "\n",
    "        query_species_num_OGs_list=[]\n",
    "        for OG_id, values in OGs_queries.items():\n",
    "            query_species_og= list(values.keys())\n",
    "            query_species_num_OGs_list.append(len(query_species_og)) \n",
    "\n",
    "        threshold_numSpecies_keep_OG_before_msa = np.sort(query_species_num_OGs_list)[-num_OG_keep_before_msa]\n",
    "\n",
    "        current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "        print(current_time, \"- Threshold for the number of species in OG is set to\",threshold_numSpecies_keep_OG_before_msa,\"to have around \",num_OG_keep_before_msa,\"OGs before MSA.\") \n",
    "\n",
    "        num_kept_OG =  sum([ 1  for i in query_species_num_OGs_list if i >= threshold_numSpecies_keep_OG_before_msa])\n",
    "        \n",
    "        if num_kept_OG > 3*num_OG_keep_before_msa:\n",
    "            print(current_time, \"- For threshold\",threshold_numSpecies_keep_OG_before_msa,\", the number of kept OGs is \",num_kept_OG,\",but is too high. Let's increase the threshold by one.\") \n",
    "            threshold_numSpecies_keep_OG_before_msa = min(len(query_species_names), threshold_numSpecies_keep_OG_before_msa+1)\n",
    "            num_kept_OG =  sum([ 1  for i in query_species_num_OGs_list if i >= threshold_numSpecies_keep_OG_before_msa])\n",
    "            print(current_time,\"- For threshold\",threshold_numSpecies_keep_OG_before_msa,\", the number of kept OGs is \",num_kept_OG) \n",
    "\n",
    "        if num_kept_OG < 0.1 * num_OG_keep_before_msa:\n",
    "            print(current_time, \"- For threshold\",threshold_numSpecies_keep_OG_before_msa,\",the number of kept OGs is \",num_kept_OG,\",but is too low. Let's decrease the threshold by one.\") \n",
    "\n",
    "            threshold_numSpecies_keep_OG_before_msa = max(1, threshold_numSpecies_keep_OG_before_msa-1)\n",
    "        \n",
    "        print(current_time, \"- Finally, the threshold is set to\",threshold_numSpecies_keep_OG_before_msa) \n",
    "\n",
    "        for OG_id, values in OGs_queries.items():\n",
    "            query_species_og= list(values.keys())\n",
    "            query_species_num_OGs_list.append(len(query_species_og)) \n",
    "            if len(query_species_og) >= threshold_numSpecies_keep_OG_before_msa:\n",
    "                OGs_queries_filtr[OG_id]=values\n",
    "                query_species_og_kept_set= query_species_og_kept_set.union(set(query_species_og))\n",
    "\n",
    "        current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "        print(current_time, \"- Finished filetering. From \",len(OGs_queries),\", now\",len(OGs_queries_filtr),\"OGs are remained.\") \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    not_included_species =[]\n",
    "    for i in set(query_species_names):\n",
    "        if i not in query_species_og_kept_set:\n",
    "            not_included_species.append(i)\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- These species missed due to the threshold of\",threshold_numSpecies_keep_OG_before_msa,\":\",not_included_species) \n",
    "    \n",
    "    return OGs_queries_filtr\n",
    "    \n",
    "    \n",
    "\n",
    "def find_best_spilce(query_species_names, prots_record_allspecies ,protnamess_splices_prots_allspecies, hogmap_allspecies ):\n",
    "\n",
    "  \n",
    "\n",
    "    (prots_hogmap_name_allspecies, prots_hogmap_hogid_allspecies, prots_hogmap_subfscore_allspecies, \n",
    "     prots_hogmap_seqlen_allspecies, prots_hogmap_subfmedseqlen_allspecies)= hogmap_allspecies\n",
    "    # for small backbone, it should be smaller to have enough gene markers\n",
    "\n",
    "\n",
    "    best_protnames_splice_prots_allspecies = []\n",
    "    best_hogid_splice_prots_allspecies = []\n",
    "    best_subfscore_splice_prots_allspecies = []\n",
    "    protnamess_splices_prots_together_allspecies = []\n",
    "    \n",
    "    for species_idx in range(len(query_species_names)):    \n",
    "        query_species_name= query_species_names[species_idx]\n",
    "\n",
    "        prots_hogmap_name = prots_hogmap_name_allspecies[species_idx]\n",
    "        prots_hogmap_hogid = prots_hogmap_hogid_allspecies[species_idx]\n",
    "        prots_hogmap_subfscore = prots_hogmap_subfscore_allspecies[species_idx]\n",
    "        prots_hogmap_seqlen = prots_hogmap_seqlen_allspecies[species_idx]\n",
    "        prots_hogmap_subfmedseqlen  = prots_hogmap_subfmedseqlen_allspecies[species_idx]\n",
    "\n",
    "        protnamess_splices_prots = protnamess_splices_prots_allspecies[species_idx]\n",
    "        prots_record= prots_record_allspecies[species_idx]\n",
    "            \n",
    "        #print(query_species_name,\"Number of prot in fa:\",len(prots_record),\", in .hogmap:\",len(prots_hogmap_name))\n",
    "\n",
    "        best_protnames_splice_prots = []\n",
    "        best_hogid_splice_prots = []\n",
    "        best_subfscore_splice_prots = []\n",
    "        previous_prot_idx=0\n",
    "\n",
    "        for protnamess_splices_prot in protnamess_splices_prots:\n",
    "\n",
    "            diff_len_splices = []\n",
    "            for protnames in protnamess_splices_prot: \n",
    "                \n",
    "                num_search_next_transcript=100 \n",
    "                # this is for speed up if the number of transcrip is low ocmpared to num prots in proteom,\n",
    "                # increase this value to have a better speed\n",
    "\n",
    "                prots_hogmap_name_sliced = prots_hogmap_name[previous_prot_idx : previous_prot_idx+num_search_next_transcript]\n",
    "\n",
    "                if protnames in set(prots_hogmap_name_sliced):\n",
    "                    prot_idx = prots_hogmap_name_sliced.index(protnames) + previous_prot_idx\n",
    "                else:\n",
    "                    try:\n",
    "                        prot_idx = prots_hogmap_name.index(protnames)\n",
    "                    except:\n",
    "                        #print(protnames,\"in splicing not found in the hogmap\")\n",
    "                        continue\n",
    "                        \n",
    "                previous_prot_idx = prot_idx\n",
    "\n",
    "                prot_hogmap_name = prots_hogmap_name[prot_idx]\n",
    "                prot_hogmap_hogid = prots_hogmap_hogid[prot_idx]\n",
    "                prot_hogmap_subfscore = prots_hogmap_subfscore[prot_idx]\n",
    "                prot_hogmap_seqlen = prots_hogmap_seqlen[prot_idx]\n",
    "                prot_hogmap_subfmedseqlen  = prots_hogmap_subfmedseqlen[prot_idx]\n",
    "\n",
    "                if not prot_hogmap_subfscore == \"na\":\n",
    "                    diff_len_splice_subf = int(prot_hogmap_seqlen)-int(prot_hogmap_subfmedseqlen)\n",
    "                    diff_len_splices.append(abs(diff_len_splice_subf))\n",
    "\n",
    "            if len(diff_len_splices):\n",
    "                best_protnames_splice = protnamess_splices_prot[diff_len_splices.index(min(diff_len_splices))]\n",
    "\n",
    "                best_protnames_splice_prots.append(best_protnames_splice)\n",
    "                best_hogid_splice_prots.append(prot_hogmap_hogid)\n",
    "                best_subfscore_splice_prots.append(prot_hogmap_subfscore)\n",
    "\n",
    "        best_protnames_splice_prots_allspecies.append(best_protnames_splice_prots)\n",
    "        best_hogid_splice_prots_allspecies.append(best_hogid_splice_prots)\n",
    "        best_subfscore_splice_prots_allspecies.append(best_subfscore_splice_prots)\n",
    "\n",
    "        protnamess_splices_prots_together = []\n",
    "        for protnamess_splices_prot in protnamess_splices_prots:\n",
    "            protnamess_splices_prots_together+=protnamess_splices_prot\n",
    "\n",
    "        #print(\"Number of prot with splicings:\",len(protnamess_splices_prots_together),\", unique splice:\",len(best_protnames_splice_prots))\n",
    "\n",
    "        protnamess_splices_prots_together_allspecies.append(protnamess_splices_prots_together)\n",
    "\n",
    "    return (best_protnames_splice_prots_allspecies, best_hogid_splice_prots_allspecies,best_subfscore_splice_prots_allspecies, protnamess_splices_prots_together_allspecies)\n",
    "\n",
    "\n",
    "\n",
    "def filter_records_allspecies(prots_record_allspecies, hogmap_allspecies, best_spilce_allspecies, subfamily_score_tresh):\n",
    "    \n",
    "    (prots_hogmap_name_allspecies, prots_hogmap_hogid_allspecies, prots_hogmap_subfscore_allspecies, \n",
    "     prots_hogmap_seqlen_allspecies, prots_hogmap_subfmedseqlen_allspecies)= hogmap_allspecies\n",
    "\n",
    "    (best_protname_splice_prots_allspecies, best_hogid_splice_prots_allspecies,best_subfscore_splice_prots_allspecies,\n",
    "     protnames_splices_prots_together_allspecies)=  best_spilce_allspecies\n",
    "\n",
    "    prots_record_filterd_allspecies = []\n",
    "    prots_name_filterd_allspecies = []\n",
    "    prots_hogid_filterd_allspecies = [] \n",
    "\n",
    "    \n",
    "    for species_idx in range(len(query_species_names)):    \n",
    "        query_species_name= query_species_names[species_idx]\n",
    "        \n",
    "        prots_record = prots_record_allspecies[species_idx]\n",
    "        print(species_idx, query_species_name)\n",
    "\n",
    "        best_protname_splice_prots = best_protname_splice_prots_allspecies[species_idx]\n",
    "        best_hogid_splice_prots = best_hogid_splice_prots_allspecies[species_idx]\n",
    "        best_subfscore_splice_prots = best_subfscore_splice_prots_allspecies[species_idx]\n",
    "        protnames_splices_prots_together = protnames_splices_prots_together_allspecies[species_idx]\n",
    "        \n",
    "        \n",
    "        prots_hogmap_name = prots_hogmap_name_allspecies[species_idx]\n",
    "        prots_hogmap_hogid = prots_hogmap_hogid_allspecies[species_idx]\n",
    "        prots_hogmap_subfscore = prots_hogmap_subfscore_allspecies[species_idx]\n",
    "\n",
    "        prots_record_filterd = []\n",
    "        prots_name_filterd = []\n",
    "        prots_hogid_filterd = []\n",
    "\n",
    "        previous_prot_idx_nosplice=0\n",
    "        previous_prot_idx_splice=0\n",
    "\n",
    "        for prot_idx in range( len(prots_record) ):\n",
    "            prot_record = prots_record[prot_idx]\n",
    "            prot_name = prot_record.id\n",
    "\n",
    "            if prot_name not in protnames_splices_prots_together:\n",
    "                # there is no splicing for this variant in the .splice file \n",
    "                prots_hogmap_name_sliced = prots_hogmap_name[previous_prot_idx_nosplice:previous_prot_idx_nosplice+100]        \n",
    "                if prot_name in set(prots_hogmap_name_sliced):\n",
    "                    prot_idx_nosplice = previous_prot_idx_nosplice + prots_hogmap_name_sliced.index(prot_name) \n",
    "                else:\n",
    "                    try:\n",
    "                        prot_idx_nosplice = prots_hogmap_name.index(prot_name)           \n",
    "                        #print(\"b\",prot_name, previous_prot_idx_splice,prot_idx_splice)                    \n",
    "                    except:\n",
    "                        #print(prot_name,\"in splicing not found in the hogmap\")\n",
    "                        continue\n",
    "                        #print(\"b\",prot_name, previous_prot_idx_nosplice,prot_idx_nosplice)\n",
    "                #print(\"here1\")\n",
    "                previous_prot_idx_nosplice = prot_idx_nosplice\n",
    "\n",
    "                prot_hogmap_hogid = prots_hogmap_hogid[prot_idx_nosplice]\n",
    "                prot_hogmap_subfscore = prots_hogmap_subfscore[prot_idx_nosplice]\n",
    "\n",
    "                if prot_hogmap_subfscore != \"na\":\n",
    "                    if float(prot_hogmap_subfscore) > subfamily_score_tresh:\n",
    "                        if prot_hogmap_hogid not in prots_hogid_filterd:  #to avoid paralogs, remove those prot mapped on same hog\n",
    "                            prots_record_filterd.append(prot_record)\n",
    "                            prots_name_filterd.append(prot_name)\n",
    "                            prots_hogid_filterd.append(prot_hogmap_hogid)\n",
    "\n",
    "            elif prot_name in best_protname_splice_prots:\n",
    "                \n",
    "                best_protname_splice_prots_sliced= best_protname_splice_prots[previous_prot_idx_splice:previous_prot_idx_splice+100]\n",
    "                if prot_name in set(best_protname_splice_prots_sliced):        \n",
    "                    prot_idx_splice = previous_prot_idx_splice + best_protname_splice_prots_sliced.index(prot_name) \n",
    "                else:\n",
    "                    try:\n",
    "                        prot_idx_splice = best_protname_splice_prots.index(prot_name) \n",
    "                        #print(\"b\",prot_name, previous_prot_idx_splice,prot_idx_splice)                    \n",
    "                    except:\n",
    "                        #print(prot_name,\"in splicing not found in the hogmap\")\n",
    "                        continue\n",
    "                        \n",
    "                previous_prot_idx_splice = prot_idx_splice\n",
    "                \n",
    "\n",
    "                prot_hogmap_hogid     = best_hogid_splice_prots[prot_idx_splice]\n",
    "                prot_hogmap_subfscore = best_subfscore_splice_prots[prot_idx_splice]\n",
    "                if prot_hogmap_subfscore != \"na\":\n",
    "                    if float(prot_hogmap_subfscore) > subfamily_score_tresh:\n",
    "                        if prot_hogmap_hogid not in prots_hogid_filterd:\n",
    "                            prots_record_filterd.append(prot_record)\n",
    "                            prots_name_filterd.append(prot_name)\n",
    "                            prots_hogid_filterd.append(prot_hogmap_hogid)\n",
    "\n",
    "        #print(len(prots_record_filterd), len(prots_name_filterd), len(prots_hogid_filterd))\n",
    "        #print(\"Number of filtered prots:\",len(prots_record_filterd))\n",
    "\n",
    "        prots_record_filterd_allspecies.append(prots_record_filterd)\n",
    "        prots_name_filterd_allspecies.append(prots_name_filterd)\n",
    "        prots_hogid_filterd_allspecies.append(prots_hogid_filterd)\n",
    "\n",
    "    return (prots_record_filterd_allspecies, prots_name_filterd_allspecies, prots_hogid_filterd_allspecies)\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "#def combine_OG_query_filtr(OGs_queries_filtr, oma_db, threshold_least_query_sepecies_in_OG,kept_oma_species_num):\n",
    "def combine_OG_query_filtr(OGs_queries_filtr, oma_db, kept_oma_species_num):\n",
    "    \n",
    "#     ########## Combine proteins of OG with queries ##################\n",
    "#     #################################################################\n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Combining queries with OG started for\",len(OGs_queries_filtr),\"OGs.\") \n",
    "    \n",
    "    species_og_dic={}\n",
    "    \n",
    "    proteins_object_OG_dic={}\n",
    "    for OG_q in OGs_queries_filtr.keys():  # OG found in the query\n",
    "\n",
    "        dic_species_prot = OGs_queries_filtr[OG_q]\n",
    "        # if  len(dic_species_prot) > threshold_least_query_sepecies_in_OG:\n",
    "        if OG_q != -1:\n",
    "            OG_members = oma_db.oma_group_members(OG_q)\n",
    "            proteins_object_OG = [db.ProteinEntry(oma_db, pr) for pr in OG_members]  # the protein IDs of og members\n",
    "            proteins_object_OG_dic[OG_q]=proteins_object_OG\n",
    "\n",
    "            species_all_og = [ str(pr.genome.uniprot_species_code) for pr in proteins_object_OG ]\n",
    "\n",
    "            for species in species_all_og:\n",
    "                if species in species_og_dic:\n",
    "                    #species_og_dic[species].append(OG_q)                        \n",
    "                    species_og_dic[species] +=1 \n",
    "                else:\n",
    "                    #species_og_dic[species]=[OG_q]\n",
    "                    species_og_dic[species] = 1\n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Number of OMA species per OG is calculated\") \n",
    "                         \n",
    "    species_oma_list=[]\n",
    "    og_num_list = []\n",
    "    for species_oma, og_num in species_og_dic.items():\n",
    "        og_num_list.append(og_num)\n",
    "        species_oma_list.append(species_oma)\n",
    "        \n",
    "    id_species_keep = np.argsort(og_num_list)[-kept_oma_species_num:]\n",
    "    species_oma_kept = [species_oma_list[i] for i in id_species_keep]\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- These OMA species are kept\",species_oma_kept) \n",
    "\n",
    "    \n",
    "    seqRecords_OG_queries = []\n",
    "    seqRecords_all_filtr = []\n",
    "    for OG_q in OGs_queries_filtr.keys():  # OG found in the query\n",
    "\n",
    "        dic_species_prot = OGs_queries_filtr[OG_q]\n",
    "#         if len(dic_species_prot) >threshold_least_query_sepecies_in_OG:\n",
    "\n",
    "        seqRecords_query_edited_all = list(dic_species_prot.values())\n",
    "        if OG_q != -1:\n",
    "            proteins_object_OG = proteins_object_OG_dic[OG_q]\n",
    "             # covnert to biopython objects\n",
    "            seqRecords_OG = []\n",
    "            for pr in proteins_object_OG:\n",
    "                species_code = str(pr.genome.uniprot_species_code)\n",
    "                if species_code in species_oma_kept:\n",
    "                    seq_record = SeqRecord(Seq(pr.sequence),species_code,'','') \n",
    "                    seqRecords_OG.append(seq_record)\n",
    "\n",
    "            seqRecords_OG_queries =seqRecords_OG + seqRecords_query_edited_all\n",
    "            current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "            #print(current_time, \" - Length of OG\",OG_q,\"was\",len(seqRecords_OG),\",now is\",len(seqRecords_OG_queries))\n",
    "            print(current_time, \" - Combining OG\",OG_q,\" with length of \",len(seqRecords_OG),\"\\t with a query \",len(seqRecords_query_edited_all),\" is just finished.\")\n",
    "\n",
    "            seqRecords_all_filtr.append(seqRecords_OG_queries)\n",
    "\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(\"\\n\", current_time, \"- Combining queries with OG is finished! number of OGs\",len(seqRecords_all_filtr)) # \n",
    "    \n",
    "    \n",
    "    \n",
    "    open_file = open(project_folder+\"fastoma_core/_file_combined_OGs_filtr.pkl\", \"wb\")\n",
    "    pickle.dump(seqRecords_all_filtr, open_file)\n",
    "    open_file.close()\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(\"\\n\", current_time, \"- The variable seqRecords_all_filtr is saved as\", project_folder+\"fastoma_core/_file_combined_OGs_filtr.pkl\") # \n",
    "    \n",
    "    return(seqRecords_all_filtr)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def run_msa_OG(seqRecords_OG_queries):\n",
    "    ############## MSA  ##############\n",
    "    ##################################\n",
    "    #current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    #print(current_time, \"- working on new OG with length of \",len(seqRecords_OG_queries))\n",
    "\n",
    "    wrapper_mafft = mafft.Mafft(seqRecords_OG_queries,datatype=\"PROTEIN\") \n",
    "    # MAfft error: Alphabet 'U' is unknown. -> add --anysymbol argument needed to define in the sourse code\n",
    "    # workaround sed \"s/U/X/g\"\n",
    "    \n",
    "    wrapper_mafft.options.options['--retree'].set_value(1)\n",
    "\n",
    "\n",
    "    run_mafft = wrapper_mafft() # it's wrapper  storing the result  and time \n",
    "    time_taken_mafft = wrapper_mafft.elapsed_time\n",
    "\n",
    "    result_mafft = wrapper_mafft.result \n",
    "    time_taken_mafft2 = wrapper_mafft.elapsed_time\n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    #print(current_time,\"- time elapsed for MSA: \",time_taken_mafft2)\n",
    "    print(current_time,\"- MSA for an OG is just finished: \",time_taken_mafft2)\n",
    "\n",
    "    return(result_mafft)\n",
    "   \n",
    "\n",
    "\n",
    "def run_msa_OG_parallel(seqRecords_all,number_max_workers):\n",
    "        \n",
    "    iterotr_OGs = 0 \n",
    "    \n",
    "    result_mafft_all_species=[]\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Parallel msa is started for \",len(seqRecords_all),\" OGs.\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    with concurrent.futures.ProcessPoolExecutor(max_workers=number_max_workers) as executor: \n",
    "        for seqRecords_OG_queries, output_values in zip(seqRecords_all, executor.map(run_msa_OG, seqRecords_all)):\n",
    "            result_mafft_all_species.append(output_values)\n",
    "\n",
    "            \n",
    "    \n",
    "    open_file = open(project_folder+\"fastoma_core/_file_msas.pkl\", \"wb\")\n",
    "    pickle.dump(result_mafft_all_species, open_file)\n",
    "    open_file.close()\n",
    "    \n",
    "    return result_mafft_all_species\n",
    "    \n",
    "\n",
    "def filter_ogs(result_mafft_all_species,ogs_keep_number):\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Filtering MSA started.\")\n",
    "\n",
    "    density_ogs=[]\n",
    "    for msa_og in result_mafft_all_species:\n",
    "\n",
    "        gap_count_og= 0\n",
    "        all_count_og=0\n",
    "        for record in msa_og:\n",
    "            seq=str(record.seq)\n",
    "            gap_count_og += seq.count(\"-\") + seq.count(\"?\") + seq.count(\".\") +seq.count(\"~\")    \n",
    "        #gap_count_ogs.append(gap_count_og)\n",
    "        density_og=gap_count_og/(len(msa_og)*len(msa_og[0]))\n",
    "        density_ogs.append(density_og)\n",
    "        #if density_ogs> treshold_density:\n",
    "    plt.hist(density_ogs,bins=100) # , bins=10\n",
    "    #plt.show()\n",
    "    plt.savefig(project_folder+\"fastoma_core/_density_ogs.pdf\")\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- The histogram of density ogs is saved.\")\n",
    "    \n",
    "    id_og_keep = np.argsort(density_ogs)[-ogs_keep_number:]\n",
    "    result_mafft_all_species_filtr = [result_mafft_all_species[i] for i in id_og_keep]\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Filtering MSA finished, keeping\",len(result_mafft_all_species_filtr),\"out of\",len(result_mafft_all_species))       \n",
    "    \n",
    "    all_labels_raw_ra = [seq.id for aln in result_mafft_all_species_filtr for seq in aln]\n",
    "    all_labels_raw_filter = [seq.id for aln in result_mafft_all_species for seq in aln]\n",
    "    \n",
    "    print(current_time, \"- Number of unique record id, species name, was\",len(set(all_labels_raw)),\"and now is\",len(set(all_labels_raw_filter)))\n",
    "    if len(set(all_labels_raw_filter)) < len(set(all_labels_raw)) :\n",
    "        print(current_time, \"- Few of species were removed due to low OG numebr , use a higher  OG number to keep all species.\") \n",
    "    \n",
    "\n",
    "    \n",
    "    open_file = open(project_folder+\"fastoma_core/_\"+str(ogs_keep_number)+\"_file_msas_filtered.pkl\", \"wb\")\n",
    "    pickle.dump(result_mafft_all_species_filtr, open_file)\n",
    "    open_file.close()\n",
    "\n",
    "        \n",
    "    return result_mafft_all_species_filtr\n",
    "\n",
    "\n",
    "\n",
    "def concatante_alignments(result_mafft_all_species,ogs_keep_number):\n",
    "    ############## Concatante alignments  ##############\n",
    "    ####################################################\n",
    "\n",
    "    #alignments= result_maf2_all\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- MSA concatenation started\")\n",
    "    \n",
    "    \n",
    "    alignments= result_mafft_all_species\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Alignments len\",len(alignments))\n",
    "    #print([len(aln) for aln in alignments ])\n",
    "    #print([len(seq) for aln in alignments for seq in aln])\n",
    "\n",
    "    all_labels_raw = [seq.id for aln in alignments for seq in aln]\n",
    "    all_labels = set(all_labels_raw)\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    #print(current_time, \"- id labels: \",len(all_labels),len(all_labels_raw))\n",
    "    \n",
    "    # Make a dictionary to store info as we go along\n",
    "    # (defaultdict is convenient -- asking for a missing key gives back an empty list)\n",
    "    concat_buf = defaultdict(list)\n",
    "\n",
    "    # Assume all alignments have same alphabet\n",
    "    alphabet = alignments[0]._alphabet\n",
    "\n",
    "    for aln in alignments:\n",
    "        length = aln.get_alignment_length()\n",
    "        #print(\"length\",length)\n",
    "        # check if any labels are missing in the current alignment\n",
    "        these_labels = set(rec.id for rec in aln)\n",
    "        missing = all_labels - these_labels\n",
    "        #print(missing)\n",
    "        # if any are missing, create unknown data of the right length,\n",
    "        # stuff the string representation into the concat_buf dict\n",
    "        for label in missing:\n",
    "            new_seq = UnknownSeq(length, alphabet=alphabet)\n",
    "            concat_buf[label].append(str(new_seq))\n",
    "\n",
    "        # else stuff the string representation into the concat_buf dict\n",
    "        for rec in aln:\n",
    "            concat_buf[rec.id].append(str(rec.seq))\n",
    "\n",
    "    # Stitch all the substrings together using join (most efficient way),\n",
    "    # and build the Biopython data structures Seq, SeqRecord and MultipleSeqAlignment\n",
    "    msa = MultipleSeqAlignment(SeqRecord(Seq(''.join(seq_arr), alphabet=alphabet), id=label)\n",
    "                                for (label, seq_arr) in concat_buf.items())\n",
    "\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- MSA concatenation finished\")\n",
    "\n",
    "    ##########\n",
    "    # print the borders of concatatnion  in nexus format\n",
    "    # http://www.iqtree.org/doc/Advanced-Tutorial\n",
    "    #######\n",
    "    out_name_msa=project_folder+\"fastoma_core/_\"+str(ogs_keep_number)+\"_msa_concatanated.txt\"\n",
    "    handle_msa_fasta = open(out_name_msa,\"w\")\n",
    "    SeqIO.write(msa, handle_msa_fasta,\"fasta\")\n",
    "    handle_msa_fasta.close()\n",
    "\n",
    "\n",
    "    # sequences_all=[]\n",
    "    # for spec, sequences in concat_buf.items():\n",
    "    #      sequences_all.append(sequences)\n",
    "    # for i in range(1,100):\n",
    "    #     print(len(sequences_all[0][i]),len(sequences_all[2][i]) ,len(sequences_all[40][i]) )\n",
    "    ## therefore, all of them are with same length of sequences\n",
    "\n",
    "    species_0=list(concat_buf.keys())[0]\n",
    "    sequences =  concat_buf[species_0]\n",
    "    len_all = [len(i) for i in sequences] \n",
    "#     borders_in_supermatrix=[]\n",
    "#     borders_in_supermatrix.append((1,len_all[0] )) \n",
    "#     accumalte = len_all[0]\n",
    "#     for i in range(1,100):\n",
    "#         borders_in_supermatrix.append((accumalte+1,accumalte+len_all[i]))\n",
    "#         accumalte +=len_all[i]\n",
    "\n",
    "#     print(len(msa),len(msa[0]),accumalte)\n",
    "#     #check it works or not\n",
    "#     # for i in range(100):\n",
    "#     #     print(borders_in_supermatrix[i][1]-borders_in_supermatrix[i][0]+1,len_all[i])\n",
    "\n",
    "#     #borders_in_supermatrix\n",
    "    \n",
    "#     # begin sets;\n",
    "#     #     charset part1 = 1-100;\n",
    "#     #     charset part2 = 101-384;\n",
    "#     #     charpartition mine = HKY+G:part1, GTR+I+G:part2;\n",
    "#     # end;\n",
    "#     borders_out_file_add=project_folder+\"border_out_parition.nex\"\n",
    "#     borders_out_file = open(borders_out_file_add,'w')\n",
    "\n",
    "#     borders_out_file.write(\"#nexus\"+\"\\n\")\n",
    "#     borders_out_file.write(\"begin sets;\"+\"\\n\")\n",
    "\n",
    "#     for i, val in enumerate(borders_in_supermatrix):\n",
    "\n",
    "#         line=\"    charset part\"+str(i)+\" = \"+str(val[0])+\"-\"+str(val[1])+\";\"\n",
    "#         #print(line)\n",
    "#         borders_out_file.write(line+'\\n')\n",
    "\n",
    "#     borders_out_file.write(\"end;\\n\")\n",
    "#     borders_out_file.close()\n",
    "#     print(borders_out_file_add)\n",
    "    \n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- MSA concatenation has been written in the file\",out_name_msa,\"with length\", len(msa),msa.get_alignment_length()) # super matrix size\n",
    "    \n",
    "    \n",
    "    return msa\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def gather_OG_2(query_species_names, prots_hogid_filterd_allspecies, prots_name_filterd_allspecies, prots_record_filterd_allspecies):\n",
    "#(query_species_names, query_hogids_filtr_species, query_prot_names_filtr_species, query_prot_records_filtr_species):\n",
    "   \n",
    "    #(query_species_names, query_hogids_filtr_species, query_prot_names_filtr_species, query_prot_records_filtr_species):\n",
    "\n",
    "    ############ Extracting the most frequent OG  ########\n",
    "    #####################################################\n",
    "\n",
    "    #dict (oma_group_nr -> dict(species, [proteins]))\n",
    "    #Og[555] = {homo_erectus: [blabla, blublu], yellow_bird: [P52134], brown_bear: [P2121,B53223]}\n",
    "\n",
    "    OGs_queries = {}\n",
    "\n",
    "    # hog_OG_map = {}\n",
    "\n",
    "    mostFrequent_OG_list_species = []\n",
    "\n",
    "    frq_most_frequent_og_list_all = [] # for development\n",
    "    \n",
    "    for species_i in  range(len(query_species_names)):\n",
    "\n",
    "        query_species_name = query_species_names[species_i]\n",
    "        #print(\"\\n\",query_species_name)\n",
    "\n",
    "        prots_hogid_filterd = prots_hogid_filterd_allspecies[species_i]\n",
    "        prots_name_filterd = prots_name_filterd_allspecies[species_i]\n",
    "        prots_record_filterd = prots_record_filterd_allspecies[species_i]\n",
    "\n",
    "        mostFrequent_OG_list=[]\n",
    "\n",
    "        for  item_idx in range(len(prots_hogid_filterd)): #\n",
    "\n",
    "            #query_protein = query_prot_names_filtr[item_idx]\n",
    "            seqRecords_query =  prots_record_filterd[item_idx]\n",
    "            seqRecords_query_edited = SeqRecord(Seq(str(seqRecords_query.seq)), query_species_name, '', '')\n",
    "            #print(seqRecords_query_edited)\n",
    "\n",
    "            hog_id= prots_hogid_filterd[item_idx]\n",
    "\n",
    "            if not hog_id in hog_OG_map:   # Caculitng  most frq OG for the new hog\n",
    "                mostFrequent_OG= -1\n",
    "                hog_OG_map[hog_id]=mostFrequent_OG\n",
    "\n",
    "            else:  # hog_id is in hog_OG_map dic\n",
    "                #print(\"using the hog-og-map\")\n",
    "                mostFrequent_OG = hog_OG_map[hog_id]\n",
    "\n",
    "            if mostFrequent_OG in OGs_queries:\n",
    "                OGs_queries_k = OGs_queries[mostFrequent_OG]\n",
    "\n",
    "                if not query_species_name in OGs_queries_k:\n",
    "                    OGs_queries_k[query_species_name] = seqRecords_query_edited\n",
    "                    OGs_queries[mostFrequent_OG] = OGs_queries_k\n",
    "            else:\n",
    "                OGs_queries[mostFrequent_OG] = {query_species_name: seqRecords_query_edited} # query_protein = query_prot_names_filtr[item_idx]\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Needed HOH-OG map \",len(OGs_queries),\"are extracted from the map file.\") \n",
    "    \n",
    "    return OGs_queries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8c023767",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "program has started. The oma database address is in  /work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/primate_oma/v27_10/omamer_database/oma_path/OmaServer.h5\n",
      "09:19:22 - OMA data is parsed and its release name is: OmaStandalone; 1.0.x\n",
      "09:19:22 - There are 10 species in the OMA database.\n",
      "OMA data is parsed and its release name is: OmaStandalone; 1.0.x\n",
      "There are 10 species in the HOG-OG OMA database.\n",
      "HOG-OG map is extracted with length of  38795\n",
      "HOG-OG map is written in the file /work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/primate_oma/v27_10/omamer_database/oma_path/OmaServer.h5_hog_og_map.dic\n",
      "09:19:38 - The are 25 species in the proteome folder.\n",
      "09:19:38 - No splicing file found for these species:  ['MANLE_', 'MICMU', 'PANTR_', 'CARSF_', 'SAIBB_', 'COLAP_']\n",
      "09:19:39 - There are  25  species in the hogmap folder.\n",
      "09:19:39 - The first species CERAT  contains  46067  proteins.\n",
      "09:19:39 - The first protein of first species is  CERAT00001\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import numpy as np\n",
    "import pyoma.browser.db as db\n",
    "import concurrent.futures\n",
    "import ast\n",
    "import pickle\n",
    "\n",
    "import zoo.wrappers.aligners.mafft as mafft  # mafft should be installed beforehand\n",
    "# import zoo.wrappers.treebuilders.fasttree as fasttree\n",
    "\n",
    "from datetime import datetime\n",
    "from sys import argv\n",
    "from os import listdir\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq, UnknownSeq\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "import matplotlib   #for development \n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "import pyoma.browser.db as db\n",
    "from pyoma.browser.hoghelper import build_hog_to_og_map\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    \n",
    "    # a global variable\n",
    "    #project_folder = argv[1]\n",
    "    project_folder = \"/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/primate_oma/v27_10/\"\n",
    "    #\"/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal/mml27/\"\n",
    "    \n",
    "    oma_database_address = project_folder+\"omamer_database/oma_path/OmaServer.h5\"\n",
    "    \n",
    "    print(\"program has started. The oma database address is in \",oma_database_address)\n",
    "    \n",
    "    (oma_db, list_oma_speices) = parse_oma_db(oma_database_address)\n",
    "\n",
    "    ## if you already  have the hog-og-map, parse it: \n",
    "    # hog_og_map_address=oma_database_address+\"_hog_og_map.dic\"    \n",
    "    # (hog_OG_map)= parse_hog_og_map(hog_og_map_address)\n",
    "\n",
    "    ## if you you don't have the hog-og-map, run this: \n",
    "    hog_OG_map = HOG_OG_map(oma_database_address)\n",
    "\n",
    "\n",
    "    #  ?? we assume that the transcript file is in order. e.g. [PROCO00019,PROCO00020] is with the same order in the proteom file and hogmap file.\n",
    "        \n",
    "    (query_species_names, prots_record_allspecies) = parse_proteome(list_oma_speices)   \n",
    "    protnames_splices_prots_allspecies = parse_splice_file(query_species_names)\n",
    "\n",
    "    hogmap_allspecies = parse_hogmap_omamer(query_species_names)\n",
    "    #(prots_hogmap_name_allspecies, prots_hogmap_hogid_allspecies, prots_hogmap_subfscore_allspecies, prots_hogmap_seqlen_allspecies, prots_hogmap_subfmedseqlen_allspecies)\n",
    "\n",
    "\n",
    "    \n",
    "    # project folder, handling_splicing,   number_max_workers\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f67b4f9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c6118fc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 CERAT\n",
      "1 MACNE_\n",
      "2 OTOGA_\n",
      "3 CHLSB\n",
      "4 PAPAN\n",
      "5 GORGO_\n",
      "6 PANPA\n",
      "7 BOVIN_\n",
      "8 MOUSE_\n",
      "9 MACFA\n",
      "10 MANLE_\n",
      "11 MICMU\n",
      "12 RHIBE\n",
      "13 PANTR_\n",
      "14 CARSF_\n",
      "15 AOTNA\n",
      "16 PONAB\n",
      "17 SAIBB_\n",
      "18 PROCO\n",
      "19 RHIRO\n",
      "20 HUMAN\n",
      "21 MACMU\n",
      "22 CALJA\n",
      "23 COLAP_\n",
      "24 NOMLE\n",
      "15:02:49 - Needed HOH-OG map  20967 are extracted from the map file.\n"
     ]
    }
   ],
   "source": [
    "if 1: \n",
    "\n",
    "    handling_splicing =  1# int(argv[2]) # 1 \n",
    "\n",
    "\n",
    "    if handling_splicing == 1:\n",
    "        best_spilce_allspecies = find_best_spilce(query_species_names, prots_record_allspecies ,protnames_splices_prots_allspecies, hogmap_allspecies )\n",
    "        # (best_protname_splice_prots_allspecies, best_hogid_splice_prots_allspecies,best_subfscore_splice_prots_allspecies, protnames_splices_prots_together_allspecies)        \n",
    "        (best_protnames_splice_prots_allspecies, best_hogid_splice_prots_allspecies, best_subfscore_splice_prots_allspecies,\n",
    "         protnames_splices_prots_together_allspecies)=  best_spilce_allspecies\n",
    "        \n",
    "        \n",
    "        # the definition of subfamily_score, It seems that I understand it wrongly and should not be used \n",
    "        \n",
    "        subfamily_score_tresh =0\n",
    "\n",
    "        (prots_record_filterd_allspecies, prots_name_filterd_allspecies, prots_hogid_filterd_allspecies) = filter_records_allspecies(prots_record_allspecies, hogmap_allspecies, best_spilce_allspecies, subfamily_score_tresh)\n",
    "\n",
    "        OGs_queries = gather_OG_2(query_species_names, prots_hogid_filterd_allspecies, prots_name_filterd_allspecies, prots_record_filterd_allspecies)\n",
    "\n",
    "    if handling_splicing == 0:\n",
    "\n",
    "\n",
    "        query_prot_records_species = prots_record_allspecies\n",
    "        query_prot_names_species_mapped = protnames_splices_prots_allspecies\n",
    "\n",
    "        (prots_hogmap_name_allspecies, prots_hogmap_hogid_allspecies, prots_hogmap_subfscore_allspecies, prots_hogmap_seqlen_allspecies, prots_hogmap_subfmedseqlen_allspecies) = hogmap_allspecies \n",
    "        query_hogids_species = prots_hogmap_hogid_allspecies\n",
    "        query_prot_names_species_mapped = prots_hogmap_name_allspecies\n",
    "\n",
    "\n",
    "        query_prot_records_species_filtered =  filter_prot_mapped(query_species_names, query_prot_records_species, query_prot_names_species_mapped)\n",
    "        (query_hogids_filtr_species, query_prot_names_filtr_species, query_prot_records_filtr_species) = extract_unique_hog_pure(query_species_names,query_hogids_species, query_prot_names_species_mapped, query_prot_records_species_filtered) \n",
    "\n",
    "        OGs_queries = gather_OG(query_species_names, query_hogids_filtr_species, query_prot_names_filtr_species, query_prot_records_filtr_species)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "214d4b75",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# import time\n",
    "# start = time.time()\n",
    "# end = time.time()\n",
    "# elapsed=end - start\n",
    "# print(elapsed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9e282e62",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15:27:15 - Start filtering OGs of proteme queries with length of  20967 for  25 species.\n",
      "15:27:15 - Threshold for the number of species in OG is set to 25 to have around  500 OGs before MSA.\n",
      "15:27:15 - For threshold 25 , the number of kept OGs is  5079 ,but is too high. Let's increase the threshold by one.\n",
      "15:27:15 - For threshold 25 , the number of kept OGs is  5079\n",
      "15:27:15 - Finally, the threshold is set to 25\n",
      "15:27:15 - Finished filetering. From  20967 , now 5079 OGs are remained.\n",
      "15:27:15 - These species missed due to the threshold of 25 : []\n"
     ]
    }
   ],
   "source": [
    "\n",
    "    num_OG_keep_before_msa=500\n",
    "    OGs_queries_filtr= filter_gathered_OG_2(OGs_queries, query_species_names, num_OG_keep_before_msa)\n",
    "\n",
    "#     threshold_least_query_sepecies_in_OG =  keep_og_treshold_species_query# 15\n",
    "\n",
    "    kept_oma_species_num = 10 # 20\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ac53807",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eab93e7d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0d70ce63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14:33:56 - Combining queries with OG started for 961 OGs.\n",
      "14:34:02 - Number of OMA species per OG is calculated\n",
      "14:34:02 - These OMA species are kept ['MOUSE', 'BOVIN', 'OTOGA', 'GORGO', 'MACNE']\n",
      "14:34:02  - Combining OG 18471  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 10646  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 18034  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 9138  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 9566  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 2894  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 9473  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 2952  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 4995  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 50  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 19393  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 19062  with length of  4 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 14836  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 10765  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 4328  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 3233  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 9336  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 9985  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 3281  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 13034  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 17672  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 8779  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 18157  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 16748  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 13581  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 20283  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 9114  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 19164  with length of  4 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 16718  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 13743  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 19346  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 12966  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 17615  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 18230  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 6708  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 10819  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 17668  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 17659  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 16526  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 18214  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 11365  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 17671  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 18847  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 8706  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 8936  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 19213  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 17446  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 15500  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 9248  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 19562  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 17915  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 19080  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 16196  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 14769  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 16499  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 2  with length of  4 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 18308  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 18621  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 8774  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 13443  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 19470  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 17657  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 19262  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 19766  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 17674  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 18332  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 11001  with length of  4 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 18837  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 20255  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 18056  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 16396  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 17040  with length of  4 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 11424  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 16016  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 17237  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 19639  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 15224  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 13918  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 15934  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 10368  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 5221  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 19137  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 17380  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 13119  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 14471  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 17760  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 10673  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 20644  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 11251  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 15340  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 17165  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 11253  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 8882  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 12023  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 11985  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 18525  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 17039  with length of  3 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 16188  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 5874  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 19037  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 16720  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 16032  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 19161  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 18098  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 16912  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 12219  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 13364  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 20274  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 16542  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 7085  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 16850  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 16828  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 16909  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 4921  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 10861  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 20543  with length of  3 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 14780  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 11804  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 18586  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 19169  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 18538  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 19293  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 19029  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 4553  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 18099  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 16717  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 20222  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 19756  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 13838  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 2328  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 8190  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 15295  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 17029  with length of  3 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 7393  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 10040  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 15564  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 17920  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 9617  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 19144  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 13713  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 10206  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 12922  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 19508  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 20423  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 18208  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 10839  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 5112  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 15809  with length of  4 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 10940  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 10073  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 11605  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 12015  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 17538  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 1183  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 19096  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 20002  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 17747  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 13154  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 16603  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 9059  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 18762  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 9348  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 10178  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 7977  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 18469  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 16967  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 11993  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 17045  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 16425  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 17447  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 18498  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 6676  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 9777  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 15795  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 6264  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 16270  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 17124  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 18459  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 19878  with length of  3 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 3940  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 2305  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 8874  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 4451  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 19191  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 8503  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 18234  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 7768  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 19229  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 9358  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 9709  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 10043  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 12332  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 18494  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 17557  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 8752  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 17468  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 19539  with length of  4 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 19154  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 10493  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 10323  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 10424  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 18871  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 8338  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 7722  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 19876  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 4535  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 5895  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 9556  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 12003  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 18362  with length of  2 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 18652  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 17989  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 19114  with length of  4 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 17859  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 18421  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 15535  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 8450  with length of  3 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 13913  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 13460  with length of  4 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 17271  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 13322  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 3401  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 16992  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 12175  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 17206  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 17072  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 12928  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 15250  with length of  4 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 18143  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 18391  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 1755  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 8223  with length of  4 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 6414  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 16332  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 19431  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 18496  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:02  - Combining OG 15435  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 8517  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 13175  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:02  - Combining OG 8440  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 15520  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 18638  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 15033  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 16932  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 18753  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:02  - Combining OG 14844  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 12089  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 17591  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 14790  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 11471  with length of  3 \t with a query  23  is just finished.\n",
      "14:34:02  - Combining OG 11868  with length of  3 \t with a query  24  is just finished.\n",
      "14:34:02  - Combining OG 10036  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 2503  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 10462  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 9436  with length of  4 \t with a query  20  is just finished.\n",
      "14:34:02  - Combining OG 17322  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 18311  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 17314  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 11833  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:02  - Combining OG 1987  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:02  - Combining OG 17004  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:02  - Combining OG 7077  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:02  - Combining OG 6197  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:02  - Combining OG 14362  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 13811  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 17967  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 16656  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 19355  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 11569  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 17998  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 16412  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 17610  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 12521  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 15613  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 11301  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 14117  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 9099  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 19835  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 11996  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 12646  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 18153  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 13162  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 1335  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 20241  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 6166  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 18734  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 19386  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 18909  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 17913  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 5841  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 12356  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 17693  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 13842  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 10744  with length of  3 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 10294  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 20144  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 20297  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 15372  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 16120  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 9969  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 19119  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 15474  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 20117  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 14881  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 19772  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 18497  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 18352  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 17957  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 18416  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 18243  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 17564  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 16950  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 17705  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 2534  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 14779  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 4472  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 18774  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 15523  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 9156  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 17051  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 9662  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 19333  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 15324  with length of  4 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 15854  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 15643  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 9189  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 4780  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 7324  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 10805  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 7797  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 12249  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 18401  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 15766  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 5936  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 18176  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 11777  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 5167  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 14079  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 9613  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 6583  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 9272  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 19775  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 14903  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 18385  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 8690  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 17877  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 14035  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 5886  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 3638  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 12531  with length of  4 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 12307  with length of  4 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 17189  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 2597  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 13306  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 13997  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 11631  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 12360  with length of  4 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 18945  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 17821  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 13750  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 12715  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 2354  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 6306  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 11414  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 4725  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 4945  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 18418  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 16269  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 19117  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 20672  with length of  2 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 20812  with length of  3 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 12713  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 16937  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 13242  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 17722  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 13559  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 17777  with length of  4 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 7756  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 11767  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 15665  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 7781  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 13911  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 12665  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 11604  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 2209  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 411  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 1152  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 16024  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 18460  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 17818  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 8873  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 5291  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 17080  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 5988  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 18084  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 9750  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 18226  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 10177  with length of  3 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 18661  with length of  4 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 17931  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 4808  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 12474  with length of  4 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 13780  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 7893  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 14887  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 2591  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 11348  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 3431  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 19380  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 12096  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 10348  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 12147  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 20544  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 12330  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 15418  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 17505  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 11816  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 3975  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 19125  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 8961  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 18778  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 18369  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 18963  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 12296  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 11976  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 8197  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 10776  with length of  4 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 15874  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 19261  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 7891  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 19441  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 6342  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 18206  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 20399  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 17476  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 18609  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 16830  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 19902  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 17510  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 5882  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 18101  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 17905  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 11248  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 18381  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 16974  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 10756  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 9190  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 7897  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 6471  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 17060  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 7537  with length of  3 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 16574  with length of  3 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 8435  with length of  3 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 6244  with length of  3 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 4080  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 7356  with length of  3 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 15803  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 19294  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 8180  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 14829  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 19014  with length of  4 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 9851  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 4508  with length of  3 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 19304  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 19968  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 18786  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 9891  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 19980  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 13669  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 18690  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 19178  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 16253  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 20249  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 16448  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 17531  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 17636  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 8267  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 7545  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 12590  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 12849  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 2816  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 4329  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 18630  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 17977  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 12653  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 20433  with length of  4 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 20340  with length of  3 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 19837  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 1923  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 10352  with length of  4 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 7698  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 18156  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 12281  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 11392  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 17454  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 18598  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 15233  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 15370  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 16411  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 11974  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 16857  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 12936  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 16685  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 19618  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 20031  with length of  4 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 8902  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 3039  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 10689  with length of  4 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 13888  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 9485  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 12284  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 11230  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 16402  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 7036  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 18166  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 15917  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 5906  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 6538  with length of  4 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 15162  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 17573  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 6529  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 7454  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 10895  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 6326  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 11291  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 19299  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 18417  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 6431  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 8594  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 16789  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 17202  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 18194  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 13202  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 18197  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 2746  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 9956  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 12134  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 8015  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 14776  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 4034  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 4039  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 15814  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 8152  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 18892  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 12049  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 19680  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 15928  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 14207  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 12317  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 11269  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 14165  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 4003  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 18564  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 13408  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 13961  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 3463  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 9806  with length of  4 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 11538  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 13168  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 18674  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 17542  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 12991  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 2178  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 13510  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 8781  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 11315  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 19249  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 13355  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 16138  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 8615  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 15216  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 19278  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 16356  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 17017  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 15712  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 15679  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 14468  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 11132  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 10132  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 3781  with length of  4 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 7481  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 3338  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 4574  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 13221  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 4817  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 12564  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 12371  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 6440  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 6612  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 17979  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 8539  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 16248  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 14717  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 17939  with length of  3 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 18055  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 10442  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 8833  with length of  3 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 19430  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 8185  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 15166  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 13850  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 15182  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 16750  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 10994  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 5517  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 10626  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 19486  with length of  3 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 7325  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 10848  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 19086  with length of  4 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 20106  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 14258  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 17638  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 11637  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 10899  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 12067  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 6668  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 5599  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 16842  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 5831  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 9201  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 15189  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 13304  with length of  3 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 8165  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 10610  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 20039  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 18676  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 9497  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 6862  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 19948  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 6639  with length of  4 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 13671  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 19254  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 17096  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 4979  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 15909  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 15811  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 6555  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 5778  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 17548  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 8855  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:03  - Combining OG 8996  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 20292  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 2722  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 18390  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 11232  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 11623  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 11084  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 7970  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 3157  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 19151  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 6858  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 3906  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 16799  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 18266  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 10000  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 20295  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 12336  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 6430  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 14321  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 13549  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 9461  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 4183  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 4133  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 7532  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 15392  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 18815  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 19961  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 11304  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 10138  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 15921  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 18758  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 9602  with length of  4 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 7285  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 18259  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 3308  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 4389  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 13444  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 9298  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 13877  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:03  - Combining OG 2338  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 1618  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 15358  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 12684  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 10070  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 12200  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 17384  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 15555  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 13861  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 16826  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 4878  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 19060  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 7697  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 10180  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:03  - Combining OG 5192  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 4400  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 7832  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 13970  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:03  - Combining OG 12588  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 12819  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 15664  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 15856  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:03  - Combining OG 3050  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 6981  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 7183  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 6637  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:03  - Combining OG 19573  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:03  - Combining OG 19103  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 19855  with length of  2 \t with a query  16  is just finished.\n",
      "14:34:03  - Combining OG 19305  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 7428  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 17356  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 11602  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:03  - Combining OG 4591  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 3559  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 20431  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:03  - Combining OG 20098  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:03  - Combining OG 14627  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:04  - Combining OG 11791  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 5937  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 17318  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 14299  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:04  - Combining OG 19599  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 19574  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 15129  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 3930  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 11144  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 17755  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 3992  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 16507  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 10217  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 8932  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 6201  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:04  - Combining OG 15077  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 5787  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 9936  with length of  4 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 17836  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 795  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 18218  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 5875  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 18662  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:04  - Combining OG 10810  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 13855  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 12383  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 12784  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 1974  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:04  - Combining OG 10457  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 14494  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 11075  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 5757  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 8562  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 13349  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 18855  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 11899  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 17310  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 17358  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 17243  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 3017  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 12414  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 19655  with length of  3 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 9057  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 7180  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 18540  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 19288  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 20537  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 18328  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:04  - Combining OG 8658  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 16438  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 7506  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 20012  with length of  4 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 17094  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 12416  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 4314  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:04  - Combining OG 11467  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 4978  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 14966  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 11770  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 9707  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 17282  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:04  - Combining OG 10965  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 9173  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 16795  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 14646  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 14422  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 18479  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 8529  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 15641  with length of  3 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 12496  with length of  3 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 15644  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 12274  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 7622  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 13159  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 5803  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 13105  with length of  4 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 12246  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 13093  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 15678  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:04  - Combining OG 16636  with length of  2 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 17079  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 5502  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 14226  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 12844  with length of  4 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 16395  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 9067  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 9369  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 5268  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 4559  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:04  - Combining OG 15598  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 16036  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 9381  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:04  - Combining OG 19471  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 15552  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 19670  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 12119  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 470  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 19384  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 9914  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 11839  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 15655  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 6664  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 6375  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 12776  with length of  4 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 15325  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 7589  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 7740  with length of  4 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 7497  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 1333  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 4396  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 10067  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 5427  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 11293  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 10290  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 16632  with length of  2 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 18408  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 14806  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 13237  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 14122  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 11343  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 9660  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 7390  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 13629  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 15230  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 17035  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 16157  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:04  - Combining OG 9609  with length of  5 \t with a query  22  is just finished.\n",
      "14:34:04  - Combining OG 6050  with length of  5 \t with a query  25  is just finished.\n",
      "14:34:04  - Combining OG 15398  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 18549  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 16056  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 18723  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 8444  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 6255  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 3693  with length of  4 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 6729  with length of  2 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 14826  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 16446  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 8172  with length of  4 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 14832  with length of  3 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 19702  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 10768  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 19689  with length of  4 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 6420  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 13052  with length of  5 \t with a query  24  is just finished.\n",
      "14:34:04  - Combining OG 1215  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 6053  with length of  3 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 20136  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 10830  with length of  2 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 11823  with length of  2 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 9253  with length of  3 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 5378  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 5189  with length of  3 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 11323  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 12568  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 9922  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 17962  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 8364  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 19982  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 18290  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 14995  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 18809  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 10925  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 18748  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 20214  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 20938  with length of  3 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 16892  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 17400  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 17214  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 3475  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 11034  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:04  - Combining OG 19796  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 13903  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 14888  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:04  - Combining OG 18544  with length of  3 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 841  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 7917  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 18925  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 3885  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 5105  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 11962  with length of  3 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 5733  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 15169  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:04  - Combining OG 13658  with length of  5 \t with a query  23  is just finished.\n",
      "14:34:04  - Combining OG 16661  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 10031  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 12815  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 18651  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 12541  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 18856  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 16602  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 19409  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 8841  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 14763  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 20624  with length of  3 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 6447  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 20049  with length of  3 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 17258  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 17972  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 13635  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 19002  with length of  4 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 16353  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 16388  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 5218  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 3563  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 19860  with length of  5 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 12528  with length of  3 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 12619  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 17033  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 19697  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 11173  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 10892  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 11180  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 2587  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 9273  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 14470  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 8876  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 14246  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 16962  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 12636  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 11858  with length of  5 \t with a query  18  is just finished.\n",
      "14:34:04  - Combining OG 12068  with length of  4 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 7302  with length of  4 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 12537  with length of  5 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 12862  with length of  3 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 7900  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 5633  with length of  3 \t with a query  20  is just finished.\n",
      "14:34:04  - Combining OG 17618  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 3732  with length of  4 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 7547  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 9464  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 14650  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 19942  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 13666  with length of  5 \t with a query  19  is just finished.\n",
      "14:34:04  - Combining OG 16754  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 19569  with length of  4 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 5949  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 16667  with length of  5 \t with a query  17  is just finished.\n",
      "14:34:04  - Combining OG 18141  with length of  5 \t with a query  16  is just finished.\n",
      "14:34:04  - Combining OG 7444  with length of  4 \t with a query  21  is just finished.\n",
      "14:34:04  - Combining OG 15525  with length of  5 \t with a query  15  is just finished.\n",
      "14:34:04  - Combining OG 19923  with length of  2 \t with a query  15  is just finished.\n",
      "\n",
      " 14:34:04 - Combining queries with OG is finished! number of OGs 960\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/primate_oma/v26_5/fastoma_core/_file_combined_OGs_filtr.pkl'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_713727/1220958264.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mseqRecords_all_filtr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcombine_OG_query_filtr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mOGs_queries_filtr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moma_db\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mkept_oma_species_num\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m#    seqRecords_all_filtr = combine_OG_query_filtr(OGs_queries_filtr, oma_db, threshold_least_query_sepecies_in_OG,kept_oma_species_num)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_713727/3587945244.py\u001b[0m in \u001b[0;36mcombine_OG_query_filtr\u001b[0;34m(OGs_queries_filtr, oma_db, kept_oma_species_num)\u001b[0m\n\u001b[1;32m    756\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    757\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 758\u001b[0;31m     \u001b[0mopen_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mproject_folder\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"fastoma_core/_file_combined_OGs_filtr.pkl\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"wb\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    759\u001b[0m     \u001b[0mpickle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdump\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseqRecords_all_filtr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mopen_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    760\u001b[0m     \u001b[0mopen_file\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/primate_oma/v26_5/fastoma_core/_file_combined_OGs_filtr.pkl'"
     ]
    }
   ],
   "source": [
    "    seqRecords_all_filtr = combine_OG_query_filtr(OGs_queries_filtr, oma_db,kept_oma_species_num)\n",
    "    \n",
    "#    seqRecords_all_filtr = combine_OG_query_filtr(OGs_queries_filtr, oma_db, threshold_least_query_sepecies_in_OG,kept_oma_species_num)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b9b9fb7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "    number_max_workers = round(int(argv[3])*1.3) # 3\n",
    "    result_mafft_all_species = run_msa_OG_parallel(seqRecords_all_filtr,number_max_workers)\n",
    "\n",
    "\n",
    "    all_id=[]\n",
    "    for a in result_mafft_all_species:\n",
    "        for i in a:\n",
    "            all_id.append(i.id)\n",
    "    len(all_id),len(set(all_id))\n",
    "\n",
    "    list_removed_species=[]\n",
    "    for i in query_species_names:\n",
    "        if i not in set(all_id):\n",
    "            list_removed_species.append(i)\n",
    "\n",
    "    if len(list_removed_species):\n",
    "        print(\"These species are removed from the input proteome:\",list_removed_species)\n",
    "    else:\n",
    "        print(\"All input species are included in the MSA.\")\n",
    "\n",
    "    list_added_species=[]\n",
    "    for i in set(all_id):\n",
    "        if i not in query_species_names:\n",
    "            list_added_species.append(i)        \n",
    "    if len(list_removed_species):\n",
    "        print(\"These species are added from OMA database:\", list_added_species)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "557a4ab2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a57a30d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EEORORORO   One species is removed \n",
    "\n",
    "#project_folder = \"/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal_uniprot/mml_v4/\"\n",
    "#/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal_uniprot/mml_v4/fastoma_core_nosplice_v5\n",
    "\n",
    "#a = project_folder+ \"fastoma_core_nosplice_v5/_file_msas.pkl\"\n",
    "\n",
    "########\n",
    "\n",
    "    ogs_keep_number = 100\n",
    "    result_mafft_all_species_filtr = filter_ogs(result_mafft_all_species,ogs_keep_number)\n",
    "    msa= concatante_alignments(result_mafft_all_species_filtr, ogs_keep_number)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b31ad9f4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # for i in range(30):\n",
    "# #     print(i, msa[i].id )\n",
    "\n",
    "# seq=msa[5].seq\n",
    "\n",
    "# gap_count_og = seq.count(\"-\") + seq.count(\"?\") + seq.count(\".\") +seq.count(\"~\") \n",
    "# gap_count_og"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49a78550",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2ec23946",
   "metadata": {},
   "source": [
    "## Load pkl file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "92625d07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seq read is loaded 1181\n",
      "09:42:57 - Filtering MSA started.\n",
      "09:42:58 - The histogram of density ogs is saved.\n",
      "09:42:58 - Filtering MSA finished, keeping 100 out of 1181\n",
      "09:42:58 - MSA concatenation started\n",
      "09:42:58 - Alignments len 100\n",
      "09:42:58 - ids:  106 10138\n",
      "09:42:58 - MSA concatenation finished\n",
      "09:42:58 - MSA concatenation has been written in the file /work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal_uniprot/mml_v4/fastoma_core/_100_msa_concatanated.txt with length 106 135151\n"
     ]
    }
   ],
   "source": [
    "#a=\"/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal/mml_v4/omamer_v4_search/fastoma_out/\"\n",
    "\n",
    "project_folder = \"/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal_uniprot/mml_v4/\"\n",
    "#/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal_uniprot/mml_v4/fastoma_core_nosplice_v5\n",
    "\n",
    "a = project_folder+ \"fastoma_core_nosplice_v5/_file_msas.pkl\"\n",
    "open_file = open(a, \"rb\") #project_folder+\"_file_msas2.pkl\"\n",
    "result_mafft_all_species = pickle.load(open_file)\n",
    "open_file.close()\n",
    "print(\"seq read is loaded\", len(result_mafft_all_species))\n",
    "\n",
    "# # result_mafft_all_species[0][:]\n",
    "ogs_keep_number = 100# 500\n",
    "result_mafft_all_species_filtr = filter_ogs(result_mafft_all_species,ogs_keep_number)\n",
    "msa2= concatante_alignments(result_mafft_all_species_filtr, ogs_keep_number)\n",
    "\n",
    "\n",
    "\n",
    "# for record in msa:\n",
    "#     if record.id ==\"S0003\":\n",
    "#         a= record.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "83a9b766",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "106"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "080d622f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "58f4ce79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(106, 135151, 135151)"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(msa2),len(msa2[0]),len(msa2[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf14eae2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "2c411b1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def msa_filter_col(msa_filtered_row, tresh_ratio_gap_col):\n",
    "\n",
    "    ratio_col_all = []\n",
    "\n",
    "    length_record= len(msa_filtered_row[1])\n",
    "    num_records = len(msa_filtered_row)\n",
    "\n",
    "\n",
    "    keep_cols = []\n",
    "    for col_i in range(length_record):  # inspired by https://github.com/andreas-wilm/compbio-utils/blob/master/prune_aln_cols.py \n",
    "\n",
    "        col_values = [record.seq[col_i] for record in msa_filtered_row]\n",
    "\n",
    "        gap_count=col_values.count(\"-\") + col_values.count(\"?\") + col_values.count(\".\") +col_values.count(\"~\")\n",
    "\n",
    "        ratio_col_nongap = 1- gap_count/num_records\n",
    "        ratio_col_all.append(ratio_col_nongap)\n",
    "        if ratio_col_nongap  > tresh_ratio_gap_col:\n",
    "            keep_cols.append(col_i)\n",
    "\n",
    "\n",
    "    plt.hist(ratio_col_all,bins=100) # , bins=10\n",
    "    #plt.show()\n",
    "    plt.savefig(project_folder+\"fastoma_core/_ratio_col.pdf\")\n",
    "\n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Columns indecis extracted. Out of \", length_record,\"columns,\",len(keep_cols),\"is remained.\") \n",
    "\n",
    "    msa_filtered_row_col = []\n",
    "\n",
    "    for record in msa_filtered_row :\n",
    "        record_seq = str(record.seq)\n",
    "\n",
    "        record_seq_edited  = ''.join([record_seq[i] for i in keep_cols  ])\n",
    "        record_edited= SeqRecord(Seq(record_seq_edited), record.id, '', '')\n",
    "        msa_filtered_row_col.append(record_edited)                         \n",
    "#ogs_keep_number\n",
    "#tresh_ratio_gap_row\n",
    "    out_name_msa=project_folder+\"fastoma_core/_\"+str(\"_\")+\"_msa_concatanated_filtered_row_\"+str(\"_\")+\"_col_\"+str(tresh_ratio_gap_col)+\".txt\"\n",
    "    handle_msa_fasta = open(out_name_msa,\"w\")\n",
    "    SeqIO.write(msa_filtered_row_col, handle_msa_fasta,\"fasta\")\n",
    "    handle_msa_fasta.close()\n",
    "    \n",
    "    current_time = datetime.now().strftime(\"%H:%M:%S\")\n",
    "    print(current_time, \"- Column-wise filtering of MSA is finished\",len(msa_filtered_row_col),len(msa_filtered_row_col[0])) \n",
    "       \n",
    "    #msa_filtered_row_col = MultipleSeqAlignment(msa_filtered_row_col)\n",
    "    return msa_filtered_row_col\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "55869cc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(107, 1249527)"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(msa),len(msa[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "db16d3f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "09:38:43 - Columns indecis extracted. Out of  1249527 columns, 476182 is remained.\n",
      "09:38:46 - Column-wise filtering of MSA is finished 107 476182\n"
     ]
    }
   ],
   "source": [
    "tresh_ratio_gap_col = 0.9\n",
    "out1= msa_filter_col(msa, tresh_ratio_gap_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be945ae7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "ebcd2a6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "09:45:47 - Columns indecis extracted. Out of  476182 columns, 295757 is remained.\n",
      "09:45:49 - Column-wise filtering of MSA is finished 107 295757\n"
     ]
    }
   ],
   "source": [
    "tresh_ratio_gap_col = 0.93\n",
    "out2= msa_filter_col(out1, tresh_ratio_gap_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "ece93a35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "09:46:02 - Columns indecis extracted. Out of  295757 columns, 209881 is remained.\n",
      "09:46:03 - Column-wise filtering of MSA is finished 107 209881\n"
     ]
    }
   ],
   "source": [
    "tresh_ratio_gap_col = 0.943\n",
    "out3= msa_filter_col(out2, tresh_ratio_gap_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "5fad0e37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "09:55:18 - Columns indecis extracted. Out of  295757 columns, 209881 is remained.\n",
      "09:55:20 - Column-wise filtering of MSA is finished 107 209881\n"
     ]
    }
   ],
   "source": [
    "tresh_ratio_gap_col = 0.9439\n",
    "out3= msa_filter_col(out2, tresh_ratio_gap_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "445ef52f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tresh_ratio_gap_col = 0.8\n",
    "# tresh_ratio_gap_row=\"norow\"\n",
    "# out2= msa_filter_col(out1, tresh_ratio_gap_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93df8aa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# reading pkl file \n",
    "# open_file = open(project_folder+\"_file_msas2.pkl\", \"rb\")\n",
    "# result_mafft_all_species = pickle.load(open_file)\n",
    "# open_file.close()\n",
    "# print(\"seq read is loaded\", len(seqRecords_all))\n",
    "\n",
    "#print(current_time, \"- Row-wise filtering of MSA is started.\")        \n",
    "#tresh_ratio_gap_row = 0.3\n",
    "#msa_filtered_row = msa_filter_row(msa,tresh_ratio_gap_row,query_species_names,ogs_keep_number)\n",
    "#print(current_time, \"- Column-wise filtering of MSA is started.\") \n",
    "#tresh_ratio_gap_col = 0.3\n",
    "#msa_filtered_row_col=  msa_filter_col(msa_filtered_row, tresh_ratio_gap_col,tresh_ratio_gap_row,ogs_keep_number)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36432298",
   "metadata": {},
   "outputs": [],
   "source": [
    "#oma_database_address = \"/work/FAC/FBM/DBC/cdessim2/default/smajidi1/fastoma/mammal/mml_v4/omamer_v4_database/oma_path_/OmaServer.h5\"\n",
    "#oma_db = db.Database(oma_database_address+\"\")\n",
    "\n",
    "#oma_db.tax.as_dict()\n",
    "#oma_db.tax.genomes.values()\n",
    "#oma_db.tax.newick()\n",
    "#import pyoma.browser.models as models\n",
    "# models.Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "8a06c87c",
   "metadata": {},
   "outputs": [],
   "source": [
    "msa_id= []\n",
    "for i in msa2:\n",
    "    msa_id.append(i.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "00dd1a68",
   "metadata": {},
   "outputs": [],
   "source": [
    "out3_id= []\n",
    "for i in out3:\n",
    "    out3_id.append(i.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "975f1dcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Gulo_gulo'}"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(out3_id)-set(msa_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "a6cf3fac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(106, 107)"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(msa_id), len(out3_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "baf4f604",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(106, 107)"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(msa2), len(out3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "1c313a23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(result_mafft_all_species_filtr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "13dda0b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out3_id.index('Gulo_gulo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "676c2f81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Gulo_gulo'"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out3_id[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "5923dbdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('??????????????????????????????????????????????????????...???'), id='Gulo_gulo', name='', description='', dbxrefs=[])"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out3[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "85c507c0",
   "metadata": {},
   "outputs": [
    {
     "ename": "NotImplementedError",
     "evalue": "SeqRecord comparison is deliberately not implemented. Explicitly compare the attributes of interest.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNotImplementedError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_713727/3215610233.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mout2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/work/FAC/FBM/DBC/cdessim2/default/smajidi1/software/miniconda3/lib/python3.8/site-packages/Bio/SeqRecord.py\u001b[0m in \u001b[0;36m__eq__\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m    801\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__eq__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    802\u001b[0m         \u001b[0;34m\"\"\"Define the equal-to operand (not implemented).\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 803\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mNotImplementedError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_NO_SEQRECORD_COMPARISON\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    804\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    805\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__ne__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNotImplementedError\u001b[0m: SeqRecord comparison is deliberately not implemented. Explicitly compare the attributes of interest."
     ]
    }
   ],
   "source": [
    "out2.remove(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "e471aaac",
   "metadata": {},
   "outputs": [],
   "source": [
    "out4 = MultipleSeqAlignment([out2[i]  for i in range(107) if i!=1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "9c515431",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10:17:57 - Columns indecis extracted. Out of  295757 columns, 295757 is remained.\n",
      "10:17:59 - Column-wise filtering of MSA is finished 106 295757\n"
     ]
    }
   ],
   "source": [
    "tresh_ratio_gap_col = 0.92\n",
    "out5= msa_filter_col(out4, tresh_ratio_gap_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "84b30556",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10:18:39 - Columns indecis extracted. Out of  295757 columns, 209737 is remained.\n",
      "10:18:41 - Column-wise filtering of MSA is finished 106 209737\n"
     ]
    }
   ],
   "source": [
    "tresh_ratio_gap_col = 0.945\n",
    "out6= msa_filter_col(out5, tresh_ratio_gap_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "17a643f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10:19:02 - Columns indecis extracted. Out of  295757 columns, 209737 is remained.\n",
      "10:19:04 - Column-wise filtering of MSA is finished 106 209737\n"
     ]
    }
   ],
   "source": [
    "tresh_ratio_gap_col = 0.946\n",
    "out6= msa_filter_col(out5, tresh_ratio_gap_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "74248488",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10:19:17 - Columns indecis extracted. Out of  295757 columns, 209737 is remained.\n",
      "10:19:18 - Column-wise filtering of MSA is finished 106 209737\n"
     ]
    }
   ],
   "source": [
    "tresh_ratio_gap_col = 0.95\n",
    "out6= msa_filter_col(out5, tresh_ratio_gap_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "a56f76f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10:19:41 - Columns indecis extracted. Out of  295757 columns, 50788 is remained.\n",
      "10:19:41 - Column-wise filtering of MSA is finished 106 50788\n"
     ]
    }
   ],
   "source": [
    "tresh_ratio_gap_col = 0.97\n",
    "out6= msa_filter_col(out5, tresh_ratio_gap_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "ca165ef6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10:19:55 - Columns indecis extracted. Out of  295757 columns, 121044 is remained.\n",
      "10:19:56 - Column-wise filtering of MSA is finished 106 121044\n"
     ]
    }
   ],
   "source": [
    "tresh_ratio_gap_col = 0.96\n",
    "out6= msa_filter_col(out5, tresh_ratio_gap_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea248b3d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
